// @harmoniclabs/cardano-ledger-ts/dist/tx/metadata/TxMetadatum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxMetadatumText = exports.TxMetadatumBytes = exports.TxMetadatumInt = exports.TxMetadatumList = exports.TxMetadatumMap = exports.isTxMetadatum = exports.txMetadatumFromCborObj = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var uint8array_utils_1 = require("@harmoniclabs/uint8array-utils");
var InvalidCborFormatError_1 = require("../../utils/InvalidCborFormatError.js");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
function txMetadatumFromCborObj(cObj) {
    if (cObj instanceof cbor_1.CborMap) {
        return new TxMetadatumMap(cObj.map.map(function (entry) { return ({
            k: txMetadatumFromCborObj(entry.k),
            v: txMetadatumFromCborObj(entry.v)
        }); }));
    }
    if (cObj instanceof cbor_1.CborArray) {
        return new TxMetadatumList(cObj.array.map(txMetadatumFromCborObj));
    }
    if (cObj instanceof cbor_1.CborUInt || cObj instanceof cbor_1.CborNegInt) {
        return new TxMetadatumInt(cObj.num);
    }
    if (cObj instanceof cbor_1.CborBytes) {
        return new TxMetadatumBytes(cObj.buffer);
    }
    if (cObj instanceof cbor_1.CborText) {
        return new TxMetadatumText(cObj.text);
    }
    throw new InvalidCborFormatError_1.InvalidCborFormatError("TxMetadatum");
}
exports.txMetadatumFromCborObj = txMetadatumFromCborObj;
function isTxMetadatum(something) {
    return (something instanceof TxMetadatumMap ||
        something instanceof TxMetadatumList ||
        something instanceof TxMetadatumInt ||
        something instanceof TxMetadatumBytes ||
        something instanceof TxMetadatumText);
}
exports.isTxMetadatum = isTxMetadatum;
function isTxMetadatumMapEntry(something) {
    return ((0, obj_utils_1.has_n_determined_keys)(something, 2, "k", "v") &&
        isTxMetadatum(something["k"]) &&
        isTxMetadatum(something["v"]));
}
var TxMetadatumMap = /** @class */ (function () {
    function TxMetadatumMap(map, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        if (!(map.every(isTxMetadatumMapEntry)))
            throw new Error("invalid entries for TxMetadatumMap");
        this.map = Object.freeze(map);
        /* Done: this.cboRref params */
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(map);
    }
    TxMetadatumMap.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    TxMetadatumMap.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxMetadatumMap.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        }
        return new cbor_1.CborMap(this.map.map(function (entry) {
            return {
                k: entry.k.toCborObj(),
                v: entry.v.toCborObj(),
            };
        }));
    };
    TxMetadatumMap.prototype.toJSON = function () { return this.toJson(); };
    TxMetadatumMap.prototype.toJson = function () {
        return this.map.map(function (entry) {
            return {
                k: entry.k.toJson(),
                v: entry.v.toJson(),
            };
        });
    };
    return TxMetadatumMap;
}());
exports.TxMetadatumMap = TxMetadatumMap;
var TxMetadatumList = /** @class */ (function () {
    function TxMetadatumList(map, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        if (!(map.every(isTxMetadatum)))
            throw new Error("invalid entries for TxMetadatumList");
        this.list = Object.freeze(map);
        /* Done: this.cboRref params */
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(map);
    }
    TxMetadatumList.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    TxMetadatumList.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxMetadatumList.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        }
        return new cbor_1.CborArray(this.list.map(function (_) { return _.toCborObj(); }));
    };
    TxMetadatumList.prototype.toJSON = function () { return this.toJson(); };
    TxMetadatumList.prototype.toJson = function () {
        return this.list.map(function (_) { return _.toJson(); });
    };
    return TxMetadatumList;
}());
exports.TxMetadatumList = TxMetadatumList;
var TxMetadatumInt = /** @class */ (function () {
    function TxMetadatumInt(n, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        this.n = BigInt(n);
    }
    TxMetadatumInt.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    TxMetadatumInt.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxMetadatumInt.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        }
        return this.n < BigInt(0) ? new cbor_1.CborNegInt(this.n) : new cbor_1.CborUInt(this.n);
    };
    TxMetadatumInt.prototype.toJSON = function () { return this.toJson(); };
    TxMetadatumInt.prototype.toJson = function () {
        return { int: this.n.toString() };
    };
    return TxMetadatumInt;
}());
exports.TxMetadatumInt = TxMetadatumInt;
var TxMetadatumBytes = /** @class */ (function () {
    function TxMetadatumBytes(bytes, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        this.bytes = bytes instanceof Uint8Array ? bytes : (0, uint8array_utils_1.fromHex)(bytes);
        /* Done: this.cboRref params */
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(bytes);
    }
    TxMetadatumBytes.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    TxMetadatumBytes.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxMetadatumBytes.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        }
        if (this.bytes.length > 64) {
            var chunks = [];
            for (var ptr = 0; ptr < this.bytes.length; ptr += 64) {
                chunks.push(new cbor_1.CborBytes(this.bytes.subarray(ptr, ptr + 64)));
            }
            return new cbor_1.CborArray(chunks);
        }
        return new cbor_1.CborBytes(this.bytes);
    };
    TxMetadatumBytes.prototype.toJSON = function () { return this.toJson(); };
    TxMetadatumBytes.prototype.toJson = function () {
        return { bytes: (0, uint8array_utils_1.toHex)(this.bytes) };
    };
    return TxMetadatumBytes;
}());
exports.TxMetadatumBytes = TxMetadatumBytes;
var TxMetadatumText = /** @class */ (function () {
    function TxMetadatumText(text, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        if (!(typeof text === "string"))
            throw new Error("invalid text");
        this.text = text;
        /* Done: this.cboRref params */
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(text);
    }
    TxMetadatumText.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    TxMetadatumText.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxMetadatumText.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        }
        if (this.text.length > 64) {
            var chunks = [];
            for (var ptr = 0; ptr < this.text.length; ptr += 64) {
                chunks.push(new cbor_1.CborText(this.text.slice(ptr, ptr + 64)));
            }
            return new cbor_1.CborArray(chunks);
        }
        return new cbor_1.CborText(this.text);
    };
    TxMetadatumText.prototype.toJSON = function () { return this.toJson(); };
    TxMetadatumText.prototype.toJson = function () {
        return { text: this.text };
    };
    return TxMetadatumText;
}());
exports.TxMetadatumText = TxMetadatumText;
