// @harmoniclabs/cardano-ledger-ts/dist/tx/metadata/TxMetadata.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxMetadata = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var InvalidCborFormatError_1 = require("../../utils/InvalidCborFormatError.js");
var TxMetadatum_1 = require("./TxMetadatum.js");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
var TxMetadata = /** @class */ (function () {
    function TxMetadata(metadata, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        var _metadata = {};
        Object.keys(metadata)
            .forEach(function (k) {
            return (0, obj_utils_1.defineReadOnlyProperty)(_metadata, BigInt(k).toString(), (function () {
                var v = metadata[k];
                if (!((0, TxMetadatum_1.isTxMetadatum)(v)))
                    throw new Error("metatdatum with label " + k + " was not instace of 'TxMetadatum'");
                return v;
            })());
        });
        this.metadata = _metadata;
        /* Done: this.cboRref params */
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(metadata);
    }
    TxMetadata.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    TxMetadata.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxMetadata.prototype.toCborObj = function () {
        var _this = this;
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        }
        return new cbor_1.CborMap(Object.keys(this.metadata).map(function (labelStr) {
            return {
                k: new cbor_1.CborUInt(BigInt(labelStr)),
                v: _this.metadata[labelStr].toCborObj()
            };
        }));
    };
    TxMetadata.fromCbor = function (cStr) {
        return TxMetadata.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cStr), { keepRef: true }));
    };
    TxMetadata.fromCborObj = function (cObj) {
        if (!(cObj instanceof cbor_1.CborMap))
            throw new InvalidCborFormatError_1.InvalidCborFormatError("TxMetadata");
        var meta = {};
        var len = cObj.map.length;
        for (var i = 0; i < len; i++) {
            var _a = cObj.map[i], k = _a.k, v = _a.v;
            if (!(k instanceof cbor_1.CborUInt))
                throw new InvalidCborFormatError_1.InvalidCborFormatError("TxMetadata");
            (0, obj_utils_1.defineReadOnlyProperty)(meta, k.num.toString(), (0, TxMetadatum_1.txMetadatumFromCborObj)(v));
        }
        return new TxMetadata(meta, (0, getSubCborRef_1.getSubCborRef)(cObj));
    };
    TxMetadata.prototype.toJSON = function () { return this.toJson(); };
    TxMetadata.prototype.toJson = function () {
        var e_1, _a;
        var json = {};
        var ks = Object.keys(this.metadata);
        try {
            for (var ks_1 = __values(ks), ks_1_1 = ks_1.next(); !ks_1_1.done; ks_1_1 = ks_1.next()) {
                var k = ks_1_1.value;
                (0, obj_utils_1.defineReadOnlyProperty)(json, k, this.metadata[k].toJson());
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (ks_1_1 && !ks_1_1.done && (_a = ks_1.return)) _a.call(ks_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return json;
    };
    return TxMetadata;
}());
exports.TxMetadata = TxMetadata;
