// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/handshake/messages/RefuseReason/RefuseReasonVersionMismatch.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RefuseReasonVersionMismatch = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var VersionNumber_1 = require("../../HandshakeVersionTable/VersionNumber.js");
var RefuseReasonVersionMismatch = /** @class */ (function () {
    function RefuseReasonVersionMismatch(validVersions, n2n) {
        if (n2n === void 0) { n2n = true; }
        this.isN2N = true;
        if (!(Array.isArray(validVersions) &&
            validVersions.every(VersionNumber_1.isExtendedVersionNumber)))
            throw new Error("invalid 'validVerisons' for 'RefuseReasonVersionMismatch'");
        this.validVersions = validVersions.filter(function (v, i, thisArr) { return thisArr.indexOf(v) === i; });
        this.isN2N = n2n;
    }
    RefuseReasonVersionMismatch.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    RefuseReasonVersionMismatch.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    RefuseReasonVersionMismatch.prototype.toCborObj = function (n2n) {
        var _this = this;
        if (n2n === void 0) { n2n = undefined; }
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(0),
            new cbor_1.CborArray(this.validVersions.map(function (v) {
                return new cbor_1.CborUInt((0, VersionNumber_1.adaptVersionNumberToMode)(v, _this.isN2N));
            }))
        ]);
    };
    RefuseReasonVersionMismatch.fromCbor = function (cbor, n2n) {
        if (n2n === void 0) { n2n = true; }
        return RefuseReasonVersionMismatch.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)), n2n);
    };
    RefuseReasonVersionMismatch.fromCborObj = function (cbor, n2n) {
        if (n2n === void 0) { n2n = true; }
        if (!(cbor instanceof cbor_1.CborArray))
            throw new Error("invalid CBOR for 'RefuseReasonVersionMismatch'");
        var _a = __read(cbor.array, 2), idx = _a[0], _versions = _a[1];
        if (!(idx instanceof cbor_1.CborUInt &&
            idx.num === BigInt(0)))
            throw new Error("invalid CBOR for 'RefuseReasonVersionMismatch'; invalid reason index");
        if (!(_versions instanceof cbor_1.CborArray))
            throw new Error("invalid CBOR for 'RefuseReasonVersionMismatch'; invalid versions field");
        return new RefuseReasonVersionMismatch(_versions.array.map(function (v) {
            if (!(v instanceof cbor_1.CborUInt))
                throw new Error("invalid CBOR for 'VersionNumber'");
            return (0, VersionNumber_1.VersionNumber)(v.num);
        }), n2n);
    };
    return RefuseReasonVersionMismatch;
}());
exports.RefuseReasonVersionMismatch = RefuseReasonVersionMismatch;
