// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/handshake/messages/RefuseReason/RefuseReasonRefuse.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RefuseReasonRefuse = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var VersionNumber_1 = require("../../HandshakeVersionTable/VersionNumber.js");
var RefuseReasonRefuse = /** @class */ (function () {
    function RefuseReasonRefuse(_a, n2n) {
        var version = _a.version, errorMessage = _a.errorMessage;
        if (n2n === void 0) { n2n = true; }
        this.isN2N = true;
        if (!(0, VersionNumber_1.isExtendedVersionNumber)(version))
            throw new Error("invalid 'validVerisons' for 'RefuseReasonRefuse'");
        this.version = (0, VersionNumber_1.VersionNumber)(version);
        this.errorMessage = String(errorMessage);
        this.isN2N = n2n;
    }
    RefuseReasonRefuse.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    RefuseReasonRefuse.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    RefuseReasonRefuse.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(2),
            new cbor_1.CborUInt((0, VersionNumber_1.adaptVersionNumberToMode)(this.version, this.isN2N)),
            new cbor_1.CborText(this.errorMessage)
        ]);
    };
    RefuseReasonRefuse.fromCbor = function (cbor, n2n) {
        if (n2n === void 0) { n2n = true; }
        return RefuseReasonRefuse.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)), n2n);
    };
    RefuseReasonRefuse.fromCborObj = function (cbor, n2n) {
        if (n2n === void 0) { n2n = true; }
        if (!(cbor instanceof cbor_1.CborArray))
            throw new Error("invalid CBOR for 'RefuseReasonRefuse'");
        var _a = __read(cbor.array, 3), idx = _a[0], _v = _a[1], _txt = _a[2];
        if (!(idx instanceof cbor_1.CborUInt &&
            idx.num === BigInt(2)))
            throw new Error("invalid CBOR for 'RefuseReasonRefuse'; invalid reason index");
        if (!(_v instanceof cbor_1.CborUInt &&
            (0, VersionNumber_1.isExtendedVersionNumber)(_v.num)))
            throw new Error("invalid CBOR for 'RefuseReasonRefuse'; invalid old version");
        if (!(_txt instanceof cbor_1.CborText))
            throw new Error("invalid CBOR for 'RefuseReasonRefuse'; invalid error message");
        return new RefuseReasonRefuse({
            version: (0, VersionNumber_1.VersionNumber)(_v.num),
            errorMessage: _txt.text
        }, n2n);
    };
    return RefuseReasonRefuse;
}());
exports.RefuseReasonRefuse = RefuseReasonRefuse;
