// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/handshake/messages/HandshakeQueryReply.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HandshakeQueryReply = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var HandshakeVersionTable_1 = require("../HandshakeVersionTable/HandshakeVersionTable.js");
var HandshakeQueryReply = /** @class */ (function () {
    function HandshakeQueryReply(_a, n2n) {
        var versionTable = _a.versionTable;
        if (n2n === void 0) { n2n = true; }
        this.isN2N = true;
        this.versionTable = (0, HandshakeVersionTable_1.normalizeVersionTableMap)(versionTable);
        this.isN2N = n2n;
    }
    HandshakeQueryReply.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    HandshakeQueryReply.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    HandshakeQueryReply.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(3),
            (0, HandshakeVersionTable_1.versionTableToCborObj)(this.versionTable, this.isN2N)
        ]);
    };
    HandshakeQueryReply.fromCbor = function (cbor, n2n) {
        if (n2n === void 0) { n2n = true; }
        return HandshakeQueryReply.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)), n2n);
    };
    HandshakeQueryReply.fromCborObj = function (cbor, n2n) {
        if (n2n === void 0) { n2n = true; }
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 2 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            Number(cbor.array[0].num) === 3))
            throw new Error("invalid CBOR for 'N2NVersionTable'");
        return new HandshakeQueryReply({
            versionTable: (0, HandshakeVersionTable_1.versionTableFromCborObj)(cbor.array[1], n2n)
        }, n2n);
    };
    return HandshakeQueryReply;
}());
exports.HandshakeQueryReply = HandshakeQueryReply;
