// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-state-query/messages/QryAcquired.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QryAcquired = void 0;
exports.isIQryAcquired = isIQryAcquired;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isIQryAcquired(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var QryAcquired = /** @class */ (function () {
    function QryAcquired() {
    }
    ;
    QryAcquired.prototype.toJSON = function () { return this.toJson(); };
    QryAcquired.prototype.toJson = function () { return {}; };
    QryAcquired.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    QryAcquired.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    QryAcquired.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(1)]);
    };
    QryAcquired.fromCbor = function (cbor) {
        return QryAcquired.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    QryAcquired.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(1)))
            throw new Error("invalid CBOR for 'QryAcquired");
        return new QryAcquired();
    };
    return QryAcquired;
}());
exports.QryAcquired = QryAcquired;
