// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-state-query/messages/QryDone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QryDone = void 0;
exports.isIQryDone = isIQryDone;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isIQryDone(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var QryDone = /** @class */ (function () {
    function QryDone() {
    }
    ;
    QryDone.prototype.toJSON = function () { return this.toJson(); };
    QryDone.prototype.toJson = function () { return {}; };
    QryDone.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    QryDone.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    QryDone.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(7)]);
    };
    QryDone.fromCbor = function (cbor) {
        return QryDone.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    QryDone.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(7)))
            throw new Error("invalid CBOR for 'QryDone");
        return new QryDone();
    };
    return QryDone;
}());
exports.QryDone = QryDone;
