// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-state-query/messages/QryQuery.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QryQuery = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
var QryQuery = /** @class */ (function () {
    function QryQuery(qry, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        var query = qry.query;
        if (!((0, cbor_1.isCborObj)(query)))
            throw new Error("invalid IQryQuery interface");
        this.query = query;
        // only query message to keep track of the original bytes
        // because data might actually be hashed by the client
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(qry);
    }
    ;
    QryQuery.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    QryQuery.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    QryQuery.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(3),
            this.query
        ]);
    };
    QryQuery.fromCbor = function (cbor) {
        var bytes = cbor instanceof Uint8Array ? cbor : (0, cbor_1.forceCborString)(cbor).toBuffer();
        return QryQuery.fromCborObj(cbor_1.Cbor.parse(bytes, { keepRef: true }), bytes);
    };
    QryQuery.fromCborObj = function (cbor, originlBytes) {
        if (originlBytes === void 0) { originlBytes = undefined; }
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 2 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(3)))
            throw new Error("invalid CBOR for 'QryQuery");
        var _a = __read(cbor.array, 2), _idx = _a[0], _queryCbor = _a[1];
        return new QryQuery({
            query: cbor.array[1]
        }, (0, getSubCborRef_1.getSubCborRef)(cbor, originlBytes));
    };
    return QryQuery;
}());
exports.QryQuery = QryQuery;
