// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-state-query/messages/QryRelease.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QryRelease = void 0;
exports.isIQryRelease = isIQryRelease;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isIQryRelease(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var QryRelease = /** @class */ (function () {
    function QryRelease() {
    }
    ;
    QryRelease.prototype.toJSON = function () { return this.toJson(); };
    QryRelease.prototype.toJson = function () { return {}; };
    QryRelease.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    QryRelease.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    QryRelease.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(5)]);
    };
    QryRelease.fromCbor = function (cbor) {
        return QryRelease.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    QryRelease.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(5)))
            throw new Error("invalid CBOR for 'QryRelease");
        return new QryRelease();
    };
    return QryRelease;
}());
exports.QryRelease = QryRelease;
