// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-state-query/messages/QryReAcquire.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QryReAcquire = void 0;
exports.isIQryReAcquire = isIQryReAcquire;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var ChainPoint_1 = require("../../types/ChainPoint.js");
;
function isIQryReAcquire(stuff) {
    return (0, obj_utils_1.isObject)(stuff) && (stuff.point === undefined || (0, ChainPoint_1.isIChainPoint)(stuff.point));
}
var QryReAcquire = /** @class */ (function () {
    function QryReAcquire(acq) {
        if (acq === void 0) { acq = {}; }
        acq = acq !== null && acq !== void 0 ? acq : {};
        if (!isIQryReAcquire(acq))
            throw new Error("invalid interface for 'QryReAcquire'");
        this.point = (0, ChainPoint_1.isIChainPoint)(acq.point) ? new ChainPoint_1.ChainPoint(acq.point) : undefined;
    }
    ;
    QryReAcquire.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    QryReAcquire.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    QryReAcquire.prototype.toCborObj = function () {
        var arr = [new cbor_1.CborUInt(this.point ? 6 : 9)];
        if (this.point) {
            arr.push(this.point.toCborObj());
        }
        return new cbor_1.CborArray(arr);
    };
    QryReAcquire.fromCbor = function (cbor) {
        return QryReAcquire.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    QryReAcquire.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 1 &&
            cbor.array[0] instanceof cbor_1.CborUInt))
            throw new Error("invalid CBOR for 'QryReAcquire");
        var num = Number(cbor.array[0].num);
        if (num === 6) {
            if (cbor.array.length < 2)
                throw new Error("invalid CBOR for 'QryReAcquire");
            return new QryReAcquire({
                point: ChainPoint_1.ChainPoint.fromCborObj(cbor.array[1])
            });
        }
        if (num === 9)
            return new QryReAcquire({});
        throw new Error("invalid CBOR for 'QryReAcquire'; unknown index: " + num.toString());
    };
    return QryReAcquire;
}());
exports.QryReAcquire = QryReAcquire;
