// @harmoniclabs/cardano-ledger-ts/dist/eras/conway/header/ConwayHeader.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConwayHeader = exports.isIConwayHeader = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var Kes_1 = require("../../common/Kes.js");
var ConwayHeaderBody_1 = require("./ConwayHeaderBody.js");
var getSubCborRef_1 = require("../../../utils/getSubCborRef.js");
var InvalidCborFormatError_1 = require("../../../utils/InvalidCborFormatError.js");
function isIConwayHeader(thing) {
    return (0, obj_utils_1.isObject)(thing) && ((0, ConwayHeaderBody_1.isIConwayHeaderBody)(thing.body) &&
        (0, Kes_1.isKesSignature)(thing.kesSignature));
}
exports.isIConwayHeader = isIConwayHeader;
var ConwayHeader = /** @class */ (function () {
    function ConwayHeader(hdr, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        if (!isIConwayHeader(hdr))
            throw new Error("Invalid ConwayHeader");
        this.body = new ConwayHeaderBody_1.ConwayHeaderBody(hdr.body);
        this.kesSignature = hdr.kesSignature;
    }
    ConwayHeader.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    ConwayHeader.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    /* header = [header_body, body_signature : $kes_signature] */
    ConwayHeader.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([
            this.body.toCborObj(),
            new cbor_1.CborBytes(this.kesSignature)
        ]);
    };
    ConwayHeader.fromCbor = function (cbor) {
        var bytes = cbor instanceof Uint8Array ? cbor : (0, cbor_1.forceCborString)(cbor).toBuffer();
        return ConwayHeader.fromCborObj(cbor_1.Cbor.parse(bytes, { keepRef: true }), bytes);
    };
    ConwayHeader.fromCborObj = function (cbor, _originalBytes) {
        // console.log("ConwayHeader.fromCborObj", cbor)
        if (!(cbor instanceof cbor_1.CborArray
        // && cbor.array.length >= 2
        ))
            throw new InvalidCborFormatError_1.InvalidCborFormatError("invalid cbor for ConwayHeader");
        var _a = __read(cbor.array, 2), cHdrBody = _a[0], cBodySignature = _a[1];
        // console.log("ConwayHeader cHdrBody", cHdrBody);
        // console.log("ConwayHeader cBodySignature", cBodySignature);
        if (!(cBodySignature instanceof cbor_1.CborBytes))
            throw new InvalidCborFormatError_1.InvalidCborFormatError("invalid cbor for ConwayHeader cBodySignature");
        var conwayHeader = new ConwayHeader({
            body: ConwayHeaderBody_1.ConwayHeaderBody.fromCborObj(cHdrBody),
            kesSignature: cBodySignature.bytes
        }, (0, getSubCborRef_1.getSubCborRef)(cbor, _originalBytes));
        // console.log("ConwayHeader", conwayHeader);
        return conwayHeader;
    };
    return ConwayHeader;
}());
exports.ConwayHeader = ConwayHeader;
