// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/tx-submission/messages/TxSubmitInit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxSubmitInit = void 0;
exports.isITxSubmitInit = isITxSubmitInit;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isITxSubmitInit(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var TxSubmitInit = /** @class */ (function () {
    function TxSubmitInit() {
    }
    ;
    TxSubmitInit.prototype.toJSON = function () { return this.toJson(); };
    TxSubmitInit.prototype.toJson = function () { return {}; };
    TxSubmitInit.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    TxSubmitInit.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxSubmitInit.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(6)]);
    };
    TxSubmitInit.fromCbor = function (cbor) {
        return TxSubmitInit.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    TxSubmitInit.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(6)))
            throw new Error("invalid CBOR for 'TxSubmitInit");
        return new TxSubmitInit();
    };
    return TxSubmitInit;
}());
exports.TxSubmitInit = TxSubmitInit;
