// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/tx-submission/messages/TxSubmitDone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxSubmitDone = void 0;
exports.isITxSubmitDone = isITxSubmitDone;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isITxSubmitDone(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var TxSubmitDone = /** @class */ (function () {
    function TxSubmitDone() {
    }
    ;
    TxSubmitDone.prototype.toJSON = function () { return this.toJson(); };
    TxSubmitDone.prototype.toJson = function () { return {}; };
    TxSubmitDone.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    TxSubmitDone.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxSubmitDone.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(4)]);
    };
    TxSubmitDone.fromCbor = function (cbor) {
        return TxSubmitDone.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    TxSubmitDone.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(4)))
            throw new Error("invalid CBOR for 'TxSubmitDone");
        return new TxSubmitDone();
    };
    return TxSubmitDone;
}());
exports.TxSubmitDone = TxSubmitDone;
