// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/handshake/n2c/N2CHandshakeVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.forceUpTo16N2CVersionNumber = exports.canBeUpTo16N2CVersionNumber = exports.isUpTo16N2CVersionNumber = exports.forceOldN2CVersionNumber = exports.canBeOldN2CVersionNumber = exports.isOldN2CVersionNumber = exports.forceN2CVersionNumber = exports.canBeN2CVersionNumber = exports.isN2CVersionNumber = exports.N2CHandshakeVersion = void 0;
var N2CHandshakeVersion;
(function (N2CHandshakeVersion) {
    N2CHandshakeVersion[N2CHandshakeVersion["v9"] = 32777] = "v9";
    N2CHandshakeVersion[N2CHandshakeVersion["v10"] = 32778] = "v10";
    N2CHandshakeVersion[N2CHandshakeVersion["v11"] = 32779] = "v11";
    N2CHandshakeVersion[N2CHandshakeVersion["v12"] = 32780] = "v12";
    N2CHandshakeVersion[N2CHandshakeVersion["v13"] = 32781] = "v13";
    N2CHandshakeVersion[N2CHandshakeVersion["v14"] = 32782] = "v14";
    N2CHandshakeVersion[N2CHandshakeVersion["v15"] = 32783] = "v15";
    N2CHandshakeVersion[N2CHandshakeVersion["v16"] = 32784] = "v16";
})(N2CHandshakeVersion || (exports.N2CHandshakeVersion = N2CHandshakeVersion = {}));
Object.freeze(N2CHandshakeVersion);
function isN2CVersionNumber(n) {
    return isOldN2CVersionNumber(n) || isUpTo16N2CVersionNumber(n);
}
exports.isN2CVersionNumber = isN2CVersionNumber;
function canBeN2CVersionNumber(n) {
    return canBeOldN2CVersionNumber(n) || canBeUpTo16N2CVersionNumber(n);
}
exports.canBeN2CVersionNumber = canBeN2CVersionNumber;
function forceN2CVersionNumber(n) {
    return canBeOldN2CVersionNumber(n) ? forceOldN2CVersionNumber(n) : forceUpTo16N2CVersionNumber(n);
}
exports.forceN2CVersionNumber = forceN2CVersionNumber;
function isOldN2CVersionNumber(n) {
    return (n === N2CHandshakeVersion.v9 ||
        n === N2CHandshakeVersion.v10 ||
        n === N2CHandshakeVersion.v11 ||
        n === N2CHandshakeVersion.v12 ||
        n === N2CHandshakeVersion.v13 ||
        n === N2CHandshakeVersion.v14);
}
exports.isOldN2CVersionNumber = isOldN2CVersionNumber;
function canBeOldN2CVersionNumber(n) {
    if (!(typeof n === "number" ||
        typeof n === "bigint"))
        return false;
    return isOldN2CVersionNumber(Number(n));
}
exports.canBeOldN2CVersionNumber = canBeOldN2CVersionNumber;
function forceOldN2CVersionNumber(n) {
    if (!(typeof n === "number" ||
        typeof n === "bigint"))
        throw new Error("can't derive 'OldVersionNumber'");
    n = Number(n);
    if (!isOldN2CVersionNumber(n))
        throw new Error("invalid 'OldN2CVersionNumber':" + n);
    return n;
}
exports.forceOldN2CVersionNumber = forceOldN2CVersionNumber;
function isUpTo16N2CVersionNumber(n) {
    return (n === N2CHandshakeVersion.v15 ||
        n === N2CHandshakeVersion.v16);
}
exports.isUpTo16N2CVersionNumber = isUpTo16N2CVersionNumber;
function canBeUpTo16N2CVersionNumber(n) {
    if (!(typeof n === "number" ||
        typeof n === "bigint"))
        return false;
    return isUpTo16N2CVersionNumber(Number(n));
}
exports.canBeUpTo16N2CVersionNumber = canBeUpTo16N2CVersionNumber;
function forceUpTo16N2CVersionNumber(n) {
    if (!(typeof n === "number" ||
        typeof n === "bigint"))
        throw new Error("can't derive 'OldVersionNumber'");
    n = Number(n);
    if (!isUpTo16N2CVersionNumber(n))
        throw new Error("invalid 'UpTo16N2CVersionNumber':" + n);
    return n;
}
exports.forceUpTo16N2CVersionNumber = forceUpTo16N2CVersionNumber;
