// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-tx-monitor/LocalTxMonitorClient.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxMonitorClient = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var MiniProtocol_1 = require("../../MiniProtocol/index.js");
var TxMonitorMessage_1 = require("./TxMonitorMessage.js");
var TxMonitorAcquire_1 = require("./messages/TxMonitorAcquire.js");
var TxMonitorDone_1 = require("./messages/TxMonitorDone.js");
var TxMonitorGetSizes_1 = require("./messages/TxMonitorGetSizes.js");
var TxMonitorNextTx_1 = require("./messages/TxMonitorNextTx.js");
var TxMonitorRelease_1 = require("./messages/TxMonitorRelease.js");
var TxMonitorReplyGetSizes_1 = require("./messages/TxMonitorReplyGetSizes.js");
var TxMonitorReplyHasTx_1 = require("./messages/TxMonitorReplyHasTx.js");
var uint8array_utils_1 = require("@harmoniclabs/uint8array-utils");
var messages_1 = require("./messages/index.js");
var roDescr = {
    writable: false,
    enumerable: true,
    configurable: false
};
var agencyHeader = {
    hasAgency: true,
    protocol: MiniProtocol_1.MiniProtocol.LocalTxMonitor
};
function _clearListeners(listeners, evt) {
    if (typeof evt === "string" && Array.isArray(listeners[evt])) {
        listeners[evt].length = 0;
        return;
    }
    listeners.done.length = 0;
    listeners.acquire.length = 0;
    listeners.acquired.length = 0;
    listeners.release.length = 0;
    listeners.nextTx.length = 0;
    listeners.replyNextTx.length = 0;
    listeners.hasTx.length = 0;
    listeners.replyHasTx.length = 0;
    listeners.getSizes.length = 0;
    listeners.replyGetSizes.length = 0;
}
function _hasListeners(listeners) {
    return (listeners.done.length > 0 ||
        listeners.acquire.length > 0 ||
        listeners.acquired.length > 0 ||
        listeners.release.length > 0 ||
        listeners.nextTx.length > 0 ||
        listeners.replyNextTx.length > 0 ||
        listeners.hasTx.length > 0 ||
        listeners.replyHasTx.length > 0 ||
        listeners.getSizes.length > 0 ||
        listeners.replyGetSizes.length > 0);
}
function msgToName(msg) {
    if (msg instanceof TxMonitorDone_1.TxMonitorDone)
        return "done";
    if (msg instanceof TxMonitorAcquire_1.TxMonitorAcquire)
        return "acquire";
    if (msg instanceof messages_1.TxMonitorAcquired)
        return "acquired";
    if (msg instanceof TxMonitorRelease_1.TxMonitorRelease)
        return "release";
    if (msg instanceof TxMonitorNextTx_1.TxMonitorNextTx)
        return "nextTx";
    if (msg instanceof messages_1.TxMonitorReplyNextTx)
        return "replyNextTx";
    if (msg instanceof messages_1.TxMonitorHasTx)
        return "hasTx";
    if (msg instanceof TxMonitorReplyHasTx_1.TxMonitorReplyHasTx)
        return "replyHasTx";
    if (msg instanceof TxMonitorGetSizes_1.TxMonitorGetSizes)
        return "getSizes";
    if (msg instanceof TxMonitorReplyGetSizes_1.TxMonitorReplyGetSizes)
        return "replyGetSizes";
    return undefined;
}
var TxMonitorClient = /** @class */ (function () {
    function TxMonitorClient(multiplexer) {
        var self = this;
        var eventListeners = {
            done: [],
            acquire: [],
            acquired: [],
            release: [],
            nextTx: [],
            replyNextTx: [],
            hasTx: [],
            replyHasTx: [],
            getSizes: [],
            replyGetSizes: [],
        };
        var onceEventListeners = {
            done: [],
            acquire: [],
            acquired: [],
            release: [],
            nextTx: [],
            replyNextTx: [],
            hasTx: [],
            replyHasTx: [],
            getSizes: [],
            replyGetSizes: [],
        };
        function hasEventListeners() {
            return _hasListeners(eventListeners) || _hasListeners(onceEventListeners);
        }
        function clearListeners(evt) {
            _clearListeners(eventListeners, evt);
            _clearListeners(onceEventListeners, evt);
        }
        function addEventListenerOnce(evt, listener) {
            var listeners = onceEventListeners[evt];
            if (!Array.isArray(listeners))
                return self;
            listeners.push(listener);
            return self;
        }
        function addEventListener(evt, listener, options) {
            if (options === null || options === void 0 ? void 0 : options.once)
                return addEventListenerOnce(evt, listener);
            var listeners = eventListeners[evt];
            if (!Array.isArray(listeners))
                return self;
            listeners.push(listener);
            return self;
        }
        function removeEventListener(evt, listener) {
            var listeners = eventListeners[evt];
            if (!Array.isArray(listeners))
                return self;
            eventListeners[evt] = listeners.filter(function (fn) { return fn !== listener; });
            onceEventListeners[evt] = onceEventListeners[evt].filter(function (fn) { return fn !== listener; });
            return self;
        }
        function dispatchEvent(evt, msg) {
            var e_1, _a;
            var listeners = eventListeners[evt];
            if (!listeners)
                return;
            try {
                for (var listeners_1 = __values(listeners), listeners_1_1 = listeners_1.next(); !listeners_1_1.done; listeners_1_1 = listeners_1.next()) {
                    var cb_1 = listeners_1_1.value;
                    cb_1(msg);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (listeners_1_1 && !listeners_1_1.done && (_a = listeners_1.return)) _a.call(listeners_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            var nListeners = eventListeners[evt].length;
            listeners = onceEventListeners[evt];
            if (evt === "error" && nListeners === 0 && listeners.length === 0) {
                throw msg instanceof Error ? msg : new Error("Unhandled error: " + msg);
            }
            var cb;
            while (cb = listeners.shift())
                cb(msg);
            return true;
        }
        Object.defineProperties(this, {
            mplexer: __assign({ value: multiplexer }, roDescr),
            clearListeners: __assign({ value: clearListeners }, roDescr),
            removeAllListeners: __assign({ value: clearListeners }, roDescr),
            addEventListener: __assign({ value: addEventListener }, roDescr),
            addListener: __assign({ value: addEventListener }, roDescr),
            on: __assign({ value: addEventListener }, roDescr),
            once: __assign({ value: addEventListenerOnce }, roDescr),
            removeEventListener: __assign({ value: removeEventListener }, roDescr),
            removeListener: __assign({ value: removeEventListener }, roDescr),
            off: __assign({ value: removeEventListener }, roDescr),
            dispatchEvent: __assign({ value: dispatchEvent }, roDescr),
            emit: __assign({ value: dispatchEvent }, roDescr),
        });
        var prevBytes = undefined;
        var queque = [];
        multiplexer.on(MiniProtocol_1.MiniProtocol.LocalTxMonitor, function (chunk) {
            if (!hasEventListeners())
                return;
            var offset = -1;
            var thing;
            var msg;
            if (prevBytes) {
                var tmp = new Uint8Array(prevBytes.length + chunk.length);
                tmp.set(prevBytes, 0);
                tmp.set(chunk, prevBytes.length);
                chunk = tmp;
                prevBytes = undefined;
            }
            while (true) {
                var originalSTLimit = Error.stackTraceLimit;
                Error.stackTraceLimit = 0;
                try {
                    thing = cbor_1.Cbor.parseWithOffset(chunk);
                }
                catch (_a) {
                    Error.stackTraceLimit = originalSTLimit;
                    // assume the error is of "missing bytes";
                    prevBytes = chunk.slice();
                    break;
                }
                finally {
                    Error.stackTraceLimit = originalSTLimit;
                }
                offset = thing.offset;
                msg = (0, TxMonitorMessage_1.txMonitorMessageFromCborObj)(thing.parsed);
                queque.unshift(msg);
                if (offset < chunk.length) {
                    // reference same memory (`subarray`)
                    // ignore the parsed bytes
                    chunk = chunk.subarray(offset);
                    continue;
                }
                else // if( offset > chunk.length )
                 {
                    prevBytes = offset === chunk.length ?
                        undefined :
                        Uint8Array.prototype.slice.call(chunk);
                    break;
                }
            }
            var msgStr;
            while (msg = queque.pop()) {
                msgStr = msgToName(msg);
                if (!msgStr)
                    continue; // ingore messages not expected by the client
                dispatchEvent(msgStr, msg);
            }
        });
    }
    TxMonitorClient.prototype.acquire = function () {
        var self = this;
        return new Promise(function (resolve) {
            function handleAcquired() {
                self.removeEventListener("acquired", handleAcquired);
                resolve();
            }
            self.addEventListener("acquired", handleAcquired);
            self.mplexer.send(new TxMonitorAcquire_1.TxMonitorAcquire().toCbor().toBuffer(), agencyHeader);
        });
    };
    TxMonitorClient.prototype.done = function () {
        this.mplexer.send(new TxMonitorDone_1.TxMonitorDone().toCbor().toBuffer(), agencyHeader);
    };
    TxMonitorClient.prototype.release = function () {
        this.mplexer.send(new TxMonitorRelease_1.TxMonitorRelease().toCbor().toBuffer(), agencyHeader);
    };
    TxMonitorClient.prototype.nextTx = function () {
        var _this = this;
        var self = this;
        return new Promise(function (resolve) {
            function handleReply(msg) {
                self.removeEventListener("replyNextTx", handleReply);
                resolve(msg);
            }
            self.addEventListener("replyNextTx", handleReply);
            _this.mplexer.send(new TxMonitorNextTx_1.TxMonitorNextTx().toCbor().toBuffer(), agencyHeader);
        });
    };
    TxMonitorClient.prototype.hasTx = function (hash) {
        var _this = this;
        var self = this;
        hash = hash instanceof Uint8Array ?
            new Uint8Array(hash) :
            (0, uint8array_utils_1.fromHex)(hash);
        return new Promise(function (resolve) {
            function handleReply(msg) {
                self.removeEventListener("replyNextTx", handleReply);
                resolve(msg.hasTx);
            }
            self.addEventListener("replyNextTx", handleReply);
            _this.mplexer.send(new messages_1.TxMonitorHasTx({
                txId: hash
            }).toCbor().toBuffer(), agencyHeader);
        });
    };
    TxMonitorClient.prototype.getSizes = function () {
        var _this = this;
        var self = this;
        return new Promise(function (resolve) {
            function handleReply(msg) {
                self.removeEventListener("replyGetSizes", handleReply);
                resolve(msg);
            }
            self.addEventListener("replyGetSizes", handleReply);
            _this.mplexer.send(new TxMonitorGetSizes_1.TxMonitorGetSizes().toCbor().toBuffer(), agencyHeader);
        });
    };
    ;
    return TxMonitorClient;
}());
exports.TxMonitorClient = TxMonitorClient;
