// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-tx-submit/LocalTxSubmitClient.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalTxSubmitClient = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var MiniProtocol_1 = require("../../MiniProtocol/index.js");
var LocalTxSubmitMessage_1 = require("./LocalTxSubmitMessage.js");
var LocalTxSubmitAccept_1 = require("./messages/LocalTxSubmitAccept.js");
var LocalTxSubmitSubmit_1 = require("./messages/LocalTxSubmitSubmit.js");
var LocalTxSubmitDone_1 = require("./messages/LocalTxSubmitDone.js");
var LocalTxSubmitReject_1 = require("./messages/LocalTxSubmitReject.js");
var roDescr = {
    writable: false,
    enumerable: true,
    configurable: false
};
function _clearListeners(listeners, evt) {
    var arr = listeners[evt];
    if (Array.isArray(arr)) {
        arr.length = 0;
        return;
    }
    listeners.submit.length = 0;
    listeners.accept.length = 0;
    listeners.reject.length = 0;
    listeners.done.length = 0;
}
function _hasEventListeners(listeners) {
    return (listeners.submit.length > 0 ||
        listeners.reject.length > 0 ||
        listeners.accept.length > 0 ||
        listeners.done.length > 0);
}
function msgToName(msg) {
    if (msg instanceof LocalTxSubmitSubmit_1.LocalTxSubmitSubmit)
        return "submit";
    if (msg instanceof LocalTxSubmitReject_1.LocalTxSubmitReject)
        return "reject";
    if (msg instanceof LocalTxSubmitAccept_1.LocalTxSubmitAccept)
        return "accept";
    if (msg instanceof LocalTxSubmitDone_1.LocalTxSubmitDone)
        return "done";
    return undefined;
}
var LocalTxSubmitClient = /** @class */ (function () {
    function LocalTxSubmitClient(multiplexer) {
        var self = this;
        var eventListeners = {
            submit: [],
            reject: [],
            accept: [],
            done: [],
        };
        var onceEventListeners = {
            submit: [],
            reject: [],
            accept: [],
            done: [],
        };
        function clearListeners(evt) {
            _clearListeners(eventListeners, evt);
            _clearListeners(onceEventListeners, evt);
        }
        function hasEventListeners() {
            return _hasEventListeners(eventListeners) || _hasEventListeners(onceEventListeners);
        }
        function addEventListenerOnce(evt, listener) {
            var listeners = onceEventListeners[evt];
            if (!Array.isArray(listeners))
                return self;
            listeners.push(listener);
            return self;
        }
        function addEventListener(evt, listener, options) {
            if (options === null || options === void 0 ? void 0 : options.once)
                return addEventListenerOnce(evt, listener);
            var listeners = eventListeners[evt];
            if (!Array.isArray(listeners))
                return self;
            listeners.push(listener);
            return self;
        }
        function removeEventListener(evt, listener) {
            var listeners = eventListeners[evt];
            if (!Array.isArray(listeners))
                return self;
            eventListeners[evt] = listeners.filter(function (fn) { return fn !== listener; });
            onceEventListeners[evt] = onceEventListeners[evt].filter(function (fn) { return fn !== listener; });
            return self;
        }
        function dispatchEvent(evt, msg) {
            var e_1, _a;
            var listeners = eventListeners[evt];
            if (!listeners)
                return;
            try {
                for (var listeners_1 = __values(listeners), listeners_1_1 = listeners_1.next(); !listeners_1_1.done; listeners_1_1 = listeners_1.next()) {
                    var cb_1 = listeners_1_1.value;
                    cb_1(msg);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (listeners_1_1 && !listeners_1_1.done && (_a = listeners_1.return)) _a.call(listeners_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            listeners = onceEventListeners[evt];
            var cb;
            while (cb = listeners.shift())
                cb(msg);
            return true;
        }
        Object.defineProperties(this, {
            mplexer: __assign({ value: multiplexer }, roDescr),
            clearListeners: __assign({ value: clearListeners }, roDescr),
            removeAllListeners: __assign({ value: clearListeners }, roDescr),
            addEventListener: __assign({ value: addEventListener }, roDescr),
            addListener: __assign({ value: addEventListener }, roDescr),
            on: __assign({ value: addEventListener }, roDescr),
            once: __assign({ value: addEventListenerOnce }, roDescr),
            removeEventListener: __assign({ value: removeEventListener }, roDescr),
            removeListener: __assign({ value: removeEventListener }, roDescr),
            off: __assign({ value: removeEventListener }, roDescr),
            dispatchEvent: __assign({ value: dispatchEvent }, roDescr),
            emit: __assign({ value: dispatchEvent }, roDescr),
        });
        var prevBytes = undefined;
        var queque = [];
        multiplexer.on(MiniProtocol_1.MiniProtocol.LocalTxSubmission, function (chunk) {
            if (!hasEventListeners())
                return;
            var offset = -1;
            var thing;
            var msg;
            if (prevBytes) {
                var tmp = new Uint8Array(prevBytes.length + chunk.length);
                tmp.set(prevBytes, 0);
                tmp.set(chunk, prevBytes.length);
                chunk = tmp;
                prevBytes = undefined;
            }
            while (true) {
                try {
                    thing = cbor_1.Cbor.parseWithOffset(chunk);
                }
                catch (_a) {
                    // assume the error is of "missing bytes";
                    prevBytes = chunk.slice();
                    break;
                }
                offset = thing.offset;
                msg = (0, LocalTxSubmitMessage_1.localTxSubmitMessageFromCborObj)(thing.parsed);
                queque.unshift(msg);
                if (offset < chunk.length) {
                    // reference same memory (`subarray`)
                    // ignore the parsed bytes
                    chunk = chunk.subarray(offset);
                    continue;
                }
                else // if( offset >= chunk.length )
                 {
                    prevBytes = undefined;
                    break;
                }
            }
            var msgStr;
            while (msg = queque.pop()) {
                msgStr = msgToName(msg);
                if (!msgStr)
                    continue; // ingore messages not expected by the client
                self.dispatchEvent(msgStr, msg);
            }
        });
    }
    LocalTxSubmitClient.prototype.submit = function (txData) {
        var self = this;
        return new Promise(function (resolve) {
            function handleAccept(_msg) {
                self.removeListener("accept", handleAccept);
                self.removeListener("reject", handleReject);
                resolve({
                    ok: true,
                    // needs to be here to have same V8 shape
                    msg: undefined
                });
            }
            function handleReject(msg) {
                self.removeListener("accept", handleAccept);
                self.removeListener("reject", handleReject);
                resolve({
                    ok: false,
                    msg: msg.reason
                });
            }
            self.on("accept", handleAccept);
            self.on("reject", handleReject);
            self.mplexer.send(new LocalTxSubmitSubmit_1.LocalTxSubmitSubmit({ tx: txData }).toCbor().toBuffer(), {
                hasAgency: true,
                protocol: MiniProtocol_1.MiniProtocol.LocalTxSubmission
            });
        });
    };
    LocalTxSubmitClient.prototype.done = function () {
        this.mplexer.send(new LocalTxSubmitDone_1.LocalTxSubmitDone().toCbor().toBuffer(), {
            hasAgency: true,
            protocol: MiniProtocol_1.MiniProtocol.LocalTxSubmission
        });
    };
    return LocalTxSubmitClient;
}());
exports.LocalTxSubmitClient = LocalTxSubmitClient;
