// @harmoniclabs/cardano-ledger-ts/dist/eras/conway/governance/GovAction/GovActionParameterChange.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GovActionParameterChange = exports.isIGovActionParameterChange = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var hashes_1 = require("../../../../hashes/index.js");
var TxOutRef_1 = require("../../../common/TxOutRef.js");
var roDescr_1 = require("../../../../utils/roDescr.js");
var GovActionType_1 = require("./GovActionType.js");
var maybeData_1 = require("../../../../utils/maybeData.js");
var protocol_1 = require("../../protocol/index.js");
function isIGovActionParameterChange(stuff) {
    return (0, obj_utils_1.isObject)(stuff) && ((stuff.govActionId === undefined || (0, TxOutRef_1.isITxOutRef)(stuff.govActionId)) &&
        (0, protocol_1.isPartialConwayProtocolParameters)(stuff.protocolParamsUpdate) &&
        (stuff.policyHash === undefined || (0, hashes_1.canBeHash28)(stuff.policyHash)));
}
exports.isIGovActionParameterChange = isIGovActionParameterChange;
var GovActionParameterChange = /** @class */ (function () {
    function GovActionParameterChange(_a, cborRef) {
        var govActionId = _a.govActionId, protocolParamsUpdate = _a.protocolParamsUpdate, policyHash = _a.policyHash;
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        Object.defineProperties(this, {
            govActionType: __assign({ value: GovActionType_1.GovActionType.ParameterChange }, roDescr_1.roDescr),
            govActionId: __assign({ value: govActionId ? new TxOutRef_1.TxOutRef(govActionId) : undefined }, roDescr_1.roDescr),
            protocolParamsUpdate: __assign({ value: protocolParamsUpdate }, roDescr_1.roDescr),
            policyHash: __assign({ value: (0, hashes_1.canBeHash28)(policyHash) ? new hashes_1.Hash28(policyHash) : undefined }, roDescr_1.roDescr),
        });
    }
    GovActionParameterChange.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    GovActionParameterChange.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    GovActionParameterChange.prototype.toCborObj = function () {
        var _a, _b, _c, _d;
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.govActionType),
            (_b = (_a = this.govActionId) === null || _a === void 0 ? void 0 : _a.toCborObj()) !== null && _b !== void 0 ? _b : new cbor_1.CborSimple(null),
            (0, protocol_1.partialConwayProtocolParametersToCborObj)(this.protocolParamsUpdate),
            (_d = (_c = this.policyHash) === null || _c === void 0 ? void 0 : _c.toCborObj()) !== null && _d !== void 0 ? _d : new cbor_1.CborSimple(null)
        ]);
    };
    GovActionParameterChange.prototype.toData = function (v) {
        var _a, _b;
        v = "v3"; // only one supported so far
        return new plutus_data_1.DataConstr(0, [
            (0, maybeData_1.maybeData)((_a = this.govActionId) === null || _a === void 0 ? void 0 : _a.toData(v)),
            (0, protocol_1.partialConwayProtocolParametersToData)(this.protocolParamsUpdate),
            (0, maybeData_1.maybeData)((_b = this.policyHash) === null || _b === void 0 ? void 0 : _b.toData(v))
        ]);
    };
    return GovActionParameterChange;
}());
exports.GovActionParameterChange = GovActionParameterChange;
