// @harmoniclabs/cardano-ledger-ts/dist/eras/mary/tx/MaryAuxiliaryData.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MaryAuxiliaryData = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var crypto_1 = require("@harmoniclabs/crypto");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var hashes_1 = require("../../../hashes/index.js");
var NativeScript_1 = require("../../../script/NativeScript.js");
var Script_1 = require("../../../script/Script.js");
var common_1 = require("../../common/index.js");
var getSubCborRef_1 = require("../../../utils/getSubCborRef.js");
var InvalidCborFormatError_1 = require("../../../utils/InvalidCborFormatError.js");
function scriptArrToCbor(scripts) {
    return new cbor_1.CborArray(scripts.map(function (script) { return new cbor_1.CborBytes(script.bytes); }));
}
var MaryAuxiliaryData = /** @class */ (function () {
    function MaryAuxiliaryData(auxData, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        if (!((0, obj_utils_1.hasOwn)(auxData, "metadata")))
            throw new Error("'MaryAuxiliaryData' is missing 'metadata' field");
        var metadata = auxData.metadata, nativeScripts = auxData.nativeScripts;
        // -------------------------------- native scripts -------------------------------- //
        if (!(metadata === undefined ||
            metadata instanceof common_1.TxMetadata))
            throw new Error("'MaryAuxiliaryData' :: 'metadata' field was not instance of 'TxMetadata'");
        this.metadata = metadata;
        // -------------------------------- native scripts -------------------------------- //
        if (nativeScripts !== undefined) {
            if (!(Array.isArray(nativeScripts) &&
                nativeScripts.every(function (script) {
                    return true;
                })))
                throw new Error("invalid nativeScripts field");
            this.nativeScripts = nativeScripts === null || nativeScripts === void 0 ? void 0 : nativeScripts.map(function (nativeScript) {
                return nativeScript instanceof Script_1.Script
                    ? nativeScript :
                    new Script_1.Script({
                        scriptType: Script_1.ScriptType.NativeScript,
                        bytes: nativeScript
                    });
            });
        }
        else {
            this.nativeScripts = undefined;
        }
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(auxData);
    }
    Object.defineProperty(MaryAuxiliaryData.prototype, "hash", {
        get: function () {
            if (this._hash instanceof hashes_1.AuxiliaryDataHash)
                return this._hash;
            this._hash = new hashes_1.AuxiliaryDataHash(new Uint8Array((0, crypto_1.blake2b_256)(this.toCbor().toBuffer())));
            return this._hash;
        },
        enumerable: false,
        configurable: true
    });
    MaryAuxiliaryData.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    MaryAuxiliaryData.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    MaryAuxiliaryData.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborTag(259, new cbor_1.CborMap([
            this.metadata === undefined ? undefined :
                {
                    k: new cbor_1.CborUInt(0),
                    v: this.metadata.toCborObj()
                },
            this.nativeScripts === undefined || this.nativeScripts.length === 0 ? undefined :
                {
                    k: new cbor_1.CborUInt(1),
                    v: scriptArrToCbor(this.nativeScripts)
                }
        ].filter(function (elem) { return elem !== undefined; })));
    };
    MaryAuxiliaryData.fromCbor = function (cStr) {
        return MaryAuxiliaryData.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cStr), { keepRef: true }));
    };
    MaryAuxiliaryData.fromCborObj = function (cObj) {
        var _a;
        // shelley; metadata only
        if (cObj instanceof cbor_1.CborMap) {
            return new MaryAuxiliaryData({
                metadata: common_1.TxMetadata.fromCborObj(cObj)
            });
        }
        // shelley multi assets; metadata + native scripts
        if (cObj instanceof cbor_1.CborArray) {
            if (!(cObj.array[1] instanceof cbor_1.CborArray))
                throw new InvalidCborFormatError_1.InvalidCborFormatError("MaryAuxiliaryData");
            return new MaryAuxiliaryData({
                metadata: common_1.TxMetadata.fromCborObj(cObj.array[0]),
                nativeScripts: cObj.array[1].array.map(NativeScript_1.nativeScriptFromCborObj)
            });
        }
        if (!(cObj instanceof cbor_1.CborTag &&
            cObj.data instanceof cbor_1.CborMap
        // && cObj.data.map.length <= 2
        ))
            throw new InvalidCborFormatError_1.InvalidCborFormatError("MaryAuxiliaryData");
        var fields = new Array(2).fill(undefined);
        var _loop_1 = function (i) {
            var v = ((_a = cObj.data.map.find(function (_a) {
                var k = _a.k;
                return k instanceof cbor_1.CborUInt && Number(k.num) === i;
            })) !== null && _a !== void 0 ? _a : { v: undefined }).v;
            if (v === undefined)
                return "continue";
            fields[i] = v;
        };
        for (var i = 0; i < 2; i++) {
            _loop_1(i);
        }
        var _b = __read(fields, 2), _metadata = _b[0], _native = _b[1];
        if (_native !== undefined && !(_native instanceof cbor_1.CborArray))
            throw new InvalidCborFormatError_1.InvalidCborFormatError("AlonzoAuxiliaryData native");
        return new MaryAuxiliaryData({
            metadata: _metadata === undefined ? undefined : common_1.TxMetadata.fromCborObj(_metadata),
            nativeScripts: _native === undefined ? undefined :
                _native.array.map(function (nativeCborObj) {
                    return new Script_1.Script({
                        scriptType: Script_1.ScriptType.NativeScript,
                        bytes: cbor_1.Cbor.encode(nativeCborObj).toBuffer()
                    });
                })
        }, (0, getSubCborRef_1.getSubCborRef)(cObj));
    };
    MaryAuxiliaryData.prototype.toJSON = function () { return this.toJson(); };
    MaryAuxiliaryData.prototype.toJson = function () {
        var _a, _b;
        return {
            metadata: (_a = this.metadata) === null || _a === void 0 ? void 0 : _a.toJson(),
            nativeScripts: (_b = this.nativeScripts) === null || _b === void 0 ? void 0 : _b.map(function (s) { return s.toJson(); }),
        };
    };
    return MaryAuxiliaryData;
}());
exports.MaryAuxiliaryData = MaryAuxiliaryData;
