// @harmoniclabs/cardano-ledger-ts/dist/eras/mary/tx/MaryTxWitnessSet.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MaryTxWitnessSet = exports.isIMaryTxWitnessSet = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var script_1 = require("../../../script/index.js");
var hashes_1 = require("../../../hashes/index.js");
var common_1 = require("../../common/index.js");
var getCborSet_1 = require("../../../utils/getCborSet.js");
var getSubCborRef_1 = require("../../../utils/getSubCborRef.js");
var InvalidCborFormatError_1 = require("../../../utils/InvalidCborFormatError.js");
;
function isUndefOrCheckedArr(stuff, arrayElemCheck) {
    return (stuff === undefined || (Array.isArray(stuff) &&
        stuff.every(arrayElemCheck)));
}
function isIMaryTxWitnessSet(set) {
    if (!(0, obj_utils_1.isObject)(set))
        return false;
    var _a = set, vkeyWitnesses = _a.vkeyWitnesses, nativeScripts = _a.nativeScripts, bootstrapWitnesses = _a.bootstrapWitnesses;
    return (isUndefOrCheckedArr(vkeyWitnesses, function (vkeyWit) { return vkeyWit instanceof common_1.VKeyWitness; }) &&
        isUndefOrCheckedArr(nativeScripts, function (ns) { return ns instanceof script_1.Script && ns.type === script_1.ScriptType.NativeScript; }) &&
        isUndefOrCheckedArr(bootstrapWitnesses, function (bootWit) { return bootWit instanceof common_1.BootstrapWitness; }));
}
exports.isIMaryTxWitnessSet = isIMaryTxWitnessSet;
var MaryTxWitnessSet = /** @class */ (function () {
    function MaryTxWitnessSet(witnesses, cborRef, allRequiredSigners) {
        if (cborRef === void 0) { cborRef = undefined; }
        if (allRequiredSigners === void 0) { allRequiredSigners = undefined; }
        var _this = this;
        var _a;
        this.cborRef = cborRef;
        if (!(isIMaryTxWitnessSet(witnesses)))
            throw new Error("invalid witnesses passed");
        var defGetter = function (name, get) {
            Object.defineProperty(_this, name, {
                get: get,
                set: function () { },
                enumerable: true,
                configurable: false
            });
            /*
            definePropertyIfNotPresent(
                this, name,
                {
                    get,
                    set: () => {},
                    enumerable: true,
                    configurable: false
                }
            )
            */
        };
        function cloneArr(arr) {
            var _a;
            return (_a = arr === null || arr === void 0 ? void 0 : arr.map(function (element) { return element.clone(); })) !== null && _a !== void 0 ? _a : [];
        }
        function defGetterArr(name, elems) {
            var _elems = elems !== null && elems !== void 0 ? elems : [];
            defGetter(name, function () { return _elems.length === 0 ? undefined : cloneArr(_elems); });
        }
        var vkeyWitnesses = witnesses.vkeyWitnesses, nativeScripts = witnesses.nativeScripts, bootstrapWitnesses = witnesses.bootstrapWitnesses;
        var _vkeyWits = (_a = vkeyWitnesses === null || vkeyWitnesses === void 0 ? void 0 : vkeyWitnesses.map(function (wit) { return wit.clone(); })) !== null && _a !== void 0 ? _a : [];
        defGetterArr("vkeyWitnesses", _vkeyWits);
        defGetterArr("nativeScripts", nativeScripts);
        defGetterArr("bootstrapWitnesses", bootstrapWitnesses);
        var _reqSigs = Array.isArray(allRequiredSigners) && allRequiredSigners.every(function (reqSig) { return reqSig instanceof hashes_1.Hash28; }) ?
            allRequiredSigners.map(function (sig) { return sig.toString(); }) :
            undefined;
        var noRequiredSigs = _reqSigs === undefined;
        Object.defineProperty(this, "isComplete", {
            get: function () { return noRequiredSigs ||
                _reqSigs.every(function (sig) {
                    return _vkeyWits.some(function (wit) { return wit.vkey.hash.toString() === sig; });
                }); },
            set: function () { },
            configurable: false,
            enumerable: true
        });
        this.addVKeyWitness = function (vkeyWit) {
            _vkeyWits.push(vkeyWit.clone());
        };
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(witnesses);
    }
    MaryTxWitnessSet.prototype.toJSON = function () { return this.toJson(); };
    MaryTxWitnessSet.prototype.toJson = function () {
        var _a, _b, _c;
        return {
            vkeyWitnesses: (_a = this.vkeyWitnesses) === null || _a === void 0 ? void 0 : _a.map(function (vkWit) { return vkWit.toJson(); }),
            nativeScripts: (_b = this.nativeScripts) === null || _b === void 0 ? void 0 : _b.map(function (ns) { return ns.toJson(); }),
            bootstrapWitnesses: (_c = this.bootstrapWitnesses) === null || _c === void 0 ? void 0 : _c.map(function (bWit) { return bWit.toJson(); }),
        };
    };
    MaryTxWitnessSet.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    MaryTxWitnessSet.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    MaryTxWitnessSet.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        }
        return new cbor_1.CborMap([
            this.vkeyWitnesses === undefined ? undefined :
                {
                    k: new cbor_1.CborUInt(0),
                    v: new cbor_1.CborArray(this.vkeyWitnesses.map(function (witness) { return witness.toCborObj(); }))
                },
            this.nativeScripts === undefined ? undefined :
                {
                    k: new cbor_1.CborUInt(1),
                    v: new cbor_1.CborArray(this.nativeScripts.map(function (nativeScript) { return nativeScript instanceof script_1.Script ?
                        cbor_1.Cbor.parse(nativeScript.cbor) :
                        (0, script_1.nativeScriptToCborObj)(nativeScript); }))
                },
            this.bootstrapWitnesses === undefined ? undefined :
                {
                    k: new cbor_1.CborUInt(2),
                    v: new cbor_1.CborArray(this.bootstrapWitnesses.map(function (w) { return w.toCborObj(); }))
                }
        ]
            .filter(function (elem) { return elem !== undefined; }));
    };
    MaryTxWitnessSet.fromCbor = function (cStr) {
        return MaryTxWitnessSet.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cStr), { keepRef: true }));
    };
    MaryTxWitnessSet.fromCborObj = function (cObj) {
        var _a;
        if (!(cObj instanceof cbor_1.CborMap
        // && cObj.map.length >= 3
        ))
            throw new InvalidCborFormatError_1.InvalidCborFormatError("MaryTxWitnessSet");
        var fields = new Array(3).fill(undefined);
        var _loop_1 = function (i) {
            var v = ((_a = cObj.map.find(function (_a) {
                var k = _a.k;
                return k instanceof cbor_1.CborUInt && Number(k.num) === i;
            })) !== null && _a !== void 0 ? _a : { v: undefined }).v;
            if (v === undefined || !(0, cbor_1.isCborObj)(v))
                return "continue";
            fields[i] = v;
        };
        for (var i = 0; i < 3; i++) {
            _loop_1(i);
        }
        var _b = __read(fields, 3), _vkey = _b[0], _native = _b[1], _bootstrap = _b[2];
        if (!((_vkey === undefined || (0, getCborSet_1.isCborSet)(_vkey)) &&
            (_native === undefined || (0, getCborSet_1.isCborSet)(_native)) &&
            (_bootstrap === undefined || (0, getCborSet_1.isCborSet)(_bootstrap))))
            throw new InvalidCborFormatError_1.InvalidCborFormatError("MaryTxWitnessSet");
        return new MaryTxWitnessSet({
            vkeyWitnesses: _vkey === undefined ? undefined : (0, getCborSet_1.getCborSet)(_vkey).map(common_1.VKeyWitness.fromCborObj),
            nativeScripts: _native === undefined ? undefined :
                (0, getCborSet_1.getCborSet)(_native).map(function (nativeCborObj) {
                    return new script_1.Script({
                        scriptType: script_1.ScriptType.NativeScript,
                        bytes: cbor_1.Cbor.encode(nativeCborObj).toBuffer()
                    });
                }),
            bootstrapWitnesses: _bootstrap === undefined ? undefined : (0, getCborSet_1.getCborSet)(_bootstrap).map(common_1.BootstrapWitness.fromCborObj)
        }, (0, getSubCborRef_1.getSubCborRef)(cObj));
    };
    return MaryTxWitnessSet;
}());
exports.MaryTxWitnessSet = MaryTxWitnessSet;
