// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-tx-monitor/messages/TxMonitorDone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxMonitorDone = void 0;
exports.isITxMonitorDone = isITxMonitorDone;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isITxMonitorDone(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var TxMonitorDone = /** @class */ (function () {
    function TxMonitorDone() {
    }
    ;
    TxMonitorDone.prototype.toJSON = function () { return this.toJson(); };
    TxMonitorDone.prototype.toJson = function () { return {}; };
    TxMonitorDone.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    TxMonitorDone.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    TxMonitorDone.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(0)]);
    };
    TxMonitorDone.fromCbor = function (cbor) {
        return TxMonitorDone.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    TxMonitorDone.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(0)))
            throw new Error("invalid CBOR for 'TxMonitorDone");
        return new TxMonitorDone();
    };
    return TxMonitorDone;
}());
exports.TxMonitorDone = TxMonitorDone;
