// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-tx-submit/messages/LocalTxSubmitAccept.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalTxSubmitAccept = void 0;
exports.isILocalTxSubmitAccept = isILocalTxSubmitAccept;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isILocalTxSubmitAccept(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var LocalTxSubmitAccept = /** @class */ (function () {
    function LocalTxSubmitAccept() {
    }
    ;
    LocalTxSubmitAccept.prototype.toJSON = function () { return this.toJson(); };
    LocalTxSubmitAccept.prototype.toJson = function () { return {}; };
    LocalTxSubmitAccept.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    LocalTxSubmitAccept.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    LocalTxSubmitAccept.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(1)]);
    };
    LocalTxSubmitAccept.fromCbor = function (cbor) {
        return LocalTxSubmitAccept.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    LocalTxSubmitAccept.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(1)))
            throw new Error("invalid CBOR for 'LocalTxSubmitAccept");
        return new LocalTxSubmitAccept();
    };
    return LocalTxSubmitAccept;
}());
exports.LocalTxSubmitAccept = LocalTxSubmitAccept;
