// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-tx-submit/messages/LocalTxSubmitSubmit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalTxSubmitSubmit = void 0;
exports.isILocalTxSubmitSubmit = isILocalTxSubmitSubmit;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
function isILocalTxSubmitSubmit(stuff) {
    return (0, obj_utils_1.isObject)(stuff) && (0, obj_utils_1.isObject)(stuff.tx) && (stuff.tx instanceof Uint8Array);
}
var LocalTxSubmitSubmit = /** @class */ (function () {
    function LocalTxSubmitSubmit(msg, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        var tx = msg.tx;
        if (!(tx instanceof Uint8Array))
            throw new Error("invalid interface for 'LocalTxSubmitSubmit'");
        this.tx = tx;
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(msg);
    }
    ;
    LocalTxSubmitSubmit.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    LocalTxSubmitSubmit.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    LocalTxSubmitSubmit.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(0),
            new cbor_1.CborBytes(this.tx)
        ]);
    };
    LocalTxSubmitSubmit.fromCbor = function (cbor) {
        var bytes = cbor instanceof Uint8Array ? cbor : (0, cbor_1.forceCborString)(cbor).toBuffer();
        return LocalTxSubmitSubmit.fromCborObj(cbor_1.Cbor.parse(bytes, { keepRef: true }), bytes);
    };
    LocalTxSubmitSubmit.fromCborObj = function (cbor, originalBytes) {
        if (originalBytes === void 0) { originalBytes = undefined; }
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 2 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(0) &&
            cbor.array[1] instanceof cbor_1.CborBytes))
            throw new Error("invalid CBOR for 'LocalTxSubmitSubmit");
        return new LocalTxSubmitSubmit({
            tx: cbor.array[1].bytes
        }, (0, getSubCborRef_1.getSubCborRef)(cbor, originalBytes));
    };
    return LocalTxSubmitSubmit;
}());
exports.LocalTxSubmitSubmit = LocalTxSubmitSubmit;
