// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/local-tx-submit/messages/LocalTxSubmitDone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalTxSubmitDone = void 0;
exports.isILocalTxSubmitDone = isILocalTxSubmitDone;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isILocalTxSubmitDone(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var LocalTxSubmitDone = /** @class */ (function () {
    function LocalTxSubmitDone() {
    }
    ;
    LocalTxSubmitDone.prototype.toJSON = function () { return this.toJson(); };
    LocalTxSubmitDone.prototype.toJson = function () { return {}; };
    LocalTxSubmitDone.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    LocalTxSubmitDone.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    LocalTxSubmitDone.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(3)]);
    };
    LocalTxSubmitDone.fromCbor = function (cbor) {
        return LocalTxSubmitDone.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    LocalTxSubmitDone.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(3)))
            throw new Error("invalid CBOR for 'LocalTxSubmitDone");
        return new LocalTxSubmitDone();
    };
    return LocalTxSubmitDone;
}());
exports.LocalTxSubmitDone = LocalTxSubmitDone;
