// @harmoniclabs/cardano-ledger-ts/dist/governance/GovAction/GovActionInitHardFork.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GovActionInitHardFork = exports.isIGovActionInitHardFork = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var tx_1 = require("../../tx/index.js");
var GovActionType_1 = require("./GovActionType.js");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var protocolVersion_1 = require("../../ledger/protocol/protocolVersion.js");
var maybeData_1 = require("../../utils/maybeData.js");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
function isIGovActionInitHardFork(stuff) {
    return (0, obj_utils_1.isObject)(stuff) && ((stuff.govActionId === undefined || (0, tx_1.isITxOutRef)(stuff.govActionId)) &&
        (0, protocolVersion_1.isIProtocolVersion)(stuff.protocolVersion));
}
exports.isIGovActionInitHardFork = isIGovActionInitHardFork;
var GovActionInitHardFork = /** @class */ (function () {
    function GovActionInitHardFork(init, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        var govActionId = init.govActionId, protocolVersion = init.protocolVersion;
        this.govActionType = GovActionType_1.GovActionType.InitHardFork;
        this.govActionId = govActionId ? new tx_1.TxOutRef(govActionId) : undefined;
        this.protocolVersion = protocolVersion;
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(init);
    }
    GovActionInitHardFork.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    GovActionInitHardFork.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    GovActionInitHardFork.prototype.toCborObj = function () {
        var _a, _b;
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.govActionType),
            (_b = (_a = this.govActionId) === null || _a === void 0 ? void 0 : _a.toCborObj()) !== null && _b !== void 0 ? _b : new cbor_1.CborSimple(null),
            new protocolVersion_1.ProtocolVersion(this.protocolVersion).toCborObj()
        ]);
    };
    GovActionInitHardFork.prototype.toData = function (v) {
        var _a;
        v = "v3"; // only one supported so far
        return new plutus_data_1.DataConstr(1, [
            (0, maybeData_1.maybeData)((_a = this.govActionId) === null || _a === void 0 ? void 0 : _a.toData(v)),
            new plutus_data_1.DataConstr(0, [
                new plutus_data_1.DataI(this.protocolVersion.major),
                new plutus_data_1.DataI(this.protocolVersion.minor)
            ])
        ]);
    };
    return GovActionInitHardFork;
}());
exports.GovActionInitHardFork = GovActionInitHardFork;
