// @harmoniclabs/cardano-ledger-ts/dist/eras/conway/governance/Constitution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Constitution = exports.isIConstitution = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var hashes_1 = require("../../../hashes/index.js");
var Anchor_1 = require("./Anchor.js");
var maybeData_1 = require("../../../utils/maybeData.js");
var getSubCborRef_1 = require("../../../utils/getSubCborRef.js");
function isIConstitution(stuff) {
    return (0, obj_utils_1.isObject)(stuff) && ((0, Anchor_1.isIAnchor)(stuff.anchor) &&
        (stuff.scriptHash === undefined || (0, hashes_1.canBeHash28)(stuff.scriptHash)));
}
exports.isIConstitution = isIConstitution;
var Constitution = /** @class */ (function () {
    function Constitution(consti, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        var anchor = consti.anchor, scriptHash = consti.scriptHash;
        this.anchor = new Anchor_1.Anchor(anchor);
        this.scriptHash = (0, hashes_1.canBeHash28)(scriptHash) ? new hashes_1.Hash28(scriptHash) : undefined;
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(consti);
    }
    Constitution.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    Constitution.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    Constitution.prototype.toCborObj = function () {
        var _a, _b;
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([
            this.anchor.toCborObj(),
            (_b = (_a = this.scriptHash) === null || _a === void 0 ? void 0 : _a.toCborObj()) !== null && _b !== void 0 ? _b : new cbor_1.CborSimple(null)
        ]);
    };
    Constitution.prototype.toData = function (v) {
        var _a;
        v = "v3"; // only one for constitution so far
        return new plutus_data_1.DataConstr(0, [
            (0, maybeData_1.maybeData)((_a = this.scriptHash) === null || _a === void 0 ? void 0 : _a.toData(v))
        ]);
    };
    return Constitution;
}());
exports.Constitution = Constitution;
