// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/block-fetch/messages/BlockFetchBatchDone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockFetchBatchDone = void 0;
exports.isIBlockFetchBatchDone = isIBlockFetchBatchDone;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isIBlockFetchBatchDone(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var BlockFetchBatchDone = /** @class */ (function () {
    function BlockFetchBatchDone() {
        this.cborRef = undefined;
    }
    ;
    BlockFetchBatchDone.prototype.toJSON = function () { return this.toJson(); };
    BlockFetchBatchDone.prototype.toJson = function () { return {}; };
    BlockFetchBatchDone.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    BlockFetchBatchDone.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    BlockFetchBatchDone.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([new cbor_1.CborUInt(5)]);
    };
    BlockFetchBatchDone.fromCbor = function (cbor) {
        return BlockFetchBatchDone.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    BlockFetchBatchDone.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(5)))
            throw new Error("invalid CBOR for 'BlockFetchBatchDone");
        return new BlockFetchBatchDone();
    };
    return BlockFetchBatchDone;
}());
exports.BlockFetchBatchDone = BlockFetchBatchDone;
