// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/block-fetch/messages/BlockFetchStartBatch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockFetchStartBatch = void 0;
exports.isIBlockFetchStartBatch = isIBlockFetchStartBatch;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isIBlockFetchStartBatch(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var BlockFetchStartBatch = /** @class */ (function () {
    function BlockFetchStartBatch() {
        this.cborRef = undefined;
    }
    ;
    BlockFetchStartBatch.prototype.toJSON = function () { return this.toJson(); };
    BlockFetchStartBatch.prototype.toJson = function () { return {}; };
    BlockFetchStartBatch.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    BlockFetchStartBatch.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    BlockFetchStartBatch.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([new cbor_1.CborUInt(2)]);
    };
    BlockFetchStartBatch.fromCbor = function (cbor) {
        return BlockFetchStartBatch.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    BlockFetchStartBatch.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(2)))
            throw new Error("invalid CBOR for 'BlockFetchStartBatch");
        return new BlockFetchStartBatch();
    };
    return BlockFetchStartBatch;
}());
exports.BlockFetchStartBatch = BlockFetchStartBatch;
