// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/block-fetch/messages/BlockFetchNoBlocks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockFetchNoBlocks = void 0;
exports.isIBlockFetchNoBlocks = isIBlockFetchNoBlocks;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isIBlockFetchNoBlocks(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var BlockFetchNoBlocks = /** @class */ (function () {
    function BlockFetchNoBlocks() {
        this.cborRef = undefined;
    }
    ;
    BlockFetchNoBlocks.prototype.toJSON = function () { return this.toJson(); };
    BlockFetchNoBlocks.prototype.toJson = function () { return {}; };
    BlockFetchNoBlocks.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    BlockFetchNoBlocks.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    BlockFetchNoBlocks.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([new cbor_1.CborUInt(3)]);
    };
    BlockFetchNoBlocks.fromCbor = function (cbor) {
        return BlockFetchNoBlocks.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    BlockFetchNoBlocks.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(3)))
            throw new Error("invalid CBOR for 'BlockFetchNoBlocks");
        return new BlockFetchNoBlocks();
    };
    return BlockFetchNoBlocks;
}());
exports.BlockFetchNoBlocks = BlockFetchNoBlocks;
