// @harmoniclabs/cardano-ledger-ts/dist/eras/babbage/protocol/BabbageProtocolParameters.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.partialBabbageProtocolParamsToJson = exports.defaultBabbageProtocolParameters = exports.partialBabbageProtocolParametersFromCborObj = exports.partialBabbageProtocolParametersToData = exports.partialBabbageProtocolParametersToCborObj = exports.isPartialBabbageProtocolParameters = exports.isBabbageProtocolParameters = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var cardano_costmodels_ts_1 = require("@harmoniclabs/cardano-costmodels-ts");
var plutus_machine_1 = require("@harmoniclabs/plutus-machine");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var ints_1 = require("../../../utils/ints.js");
var Rational_1 = require("../../common/Rational.js");
var protocolVersion_1 = require("./protocolVersion.js");
var uint8array_utils_1 = require("@harmoniclabs/uint8array-utils");
function isBabbageProtocolParameters(something) {
    var expectedKeys = [
        "txFeePerByte",
        "txFeeFixed",
        "maxBlockBodySize",
        "maxTxSize",
        "maxBlockHeaderSize",
        "stakeAddressDeposit",
        "stakePoolDeposit",
        "poolRetireMaxEpoch",
        "stakePoolTargetNum",
        "poolPledgeInfluence",
        "monetaryExpansion",
        "treasuryCut",
        "protocolVersion",
        "minPoolCost",
        "utxoCostPerByte",
        "costModels",
        "executionUnitPrices",
        "maxTxExecutionUnits",
        "maxBlockExecutionUnits",
        "maxValueSize",
        "collateralPercentage",
        "maxCollateralInputs",
    ];
    var actualKeys = Object.keys(something);
    if (!expectedKeys.every(function (k) { return actualKeys.includes(k); }))
        return false;
    var pp = something;
    if (![
        "txFeePerByte",
        "txFeeFixed",
        "maxBlockBodySize",
        "maxTxSize",
        "maxBlockHeaderSize",
        "stakeAddressDeposit",
        "stakePoolDeposit",
        "poolRetireMaxEpoch",
        "stakePoolTargetNum",
        "minPoolCost",
        "utxoCostPerByte",
        "maxValueSize",
        "collateralPercentage",
        "maxCollateralInputs",
    ].every(function (uintKey) { return (0, ints_1.canBeUInteger)(pp[uintKey]); }))
        return false;
    if (!((0, Rational_1.isRational)(pp.poolPledgeInfluence) &&
        (0, Rational_1.isRational)(pp.monetaryExpansion) &&
        (0, Rational_1.isRational)(pp.treasuryCut)))
        return false;
    var ppv = pp.protocolVersion;
    if (!(
    // protocolVersion removed in conway
    ppv === undefined ||
        (0, protocolVersion_1.isIProtocolVersion)(ppv)))
        return false;
    var ppexecCosts = pp.executionUnitPrices;
    if (!((Array.isArray(ppexecCosts) &&
        ppexecCosts.length >= 2 &&
        ppexecCosts[0] instanceof cbor_1.CborPositiveRational &&
        ppexecCosts[1] instanceof cbor_1.CborPositiveRational) ||
        ((0, obj_utils_1.isObject)(ppexecCosts) &&
            typeof ppexecCosts.priceSteps === "number" &&
            typeof ppexecCosts.priceMemory === "number")))
        return false;
    if (!(pp.maxTxExecutionUnits instanceof plutus_machine_1.ExBudget || plutus_machine_1.ExBudget.isJson(pp.maxTxExecutionUnits) &&
        pp.maxBlockExecutionUnits instanceof plutus_machine_1.ExBudget || plutus_machine_1.ExBudget.isJson(pp.maxBlockExecutionUnits)))
        return false;
    if (!((0, cardano_costmodels_ts_1.isCostModels)(pp.costModels)))
        return false;
    return true;
}
exports.isBabbageProtocolParameters = isBabbageProtocolParameters;
function isPartialBabbageProtocolParameters(something) {
    if (!(0, obj_utils_1.isObject)(something))
        return false;
    var pp = something;
    if (![
        "txFeePerByte",
        "txFeeFixed",
        "maxBlockBodySize",
        "maxTxSize",
        "maxBlockHeaderSize",
        "stakeAddressDeposit",
        "stakePoolDeposit",
        "poolRetireMaxEpoch",
        "stakePoolTargetNum",
        "minPoolCost",
        "utxoCostPerByte",
        "maxValueSize",
        "collateralPercentage",
        "maxCollateralInputs",
    ].every(function (uintKey) { return pp[uintKey] === undefined || (0, ints_1.canBeUInteger)(pp[uintKey]); }))
        return false;
    if (!((0, Rational_1.isRationalOrUndefined)(pp.poolPledgeInfluence) &&
        (0, Rational_1.isRationalOrUndefined)(pp.monetaryExpansion) &&
        (0, Rational_1.isRationalOrUndefined)(pp.treasuryCut)))
        return false;
    var ppv = pp.protocolVersion;
    if (!(ppv === undefined ||
        (0, protocolVersion_1.isIProtocolVersion)(ppv)))
        return false;
    var ppexecCosts = pp.executionUnitPrices;
    if (!(ppexecCosts === undefined ||
        (Array.isArray(ppexecCosts) &&
            ppexecCosts.length >= 2 &&
            ppexecCosts[0] instanceof cbor_1.CborPositiveRational &&
            ppexecCosts[1] instanceof cbor_1.CborPositiveRational) || ((0, obj_utils_1.isObject)(ppexecCosts) &&
        typeof ppexecCosts.priceMemory === "number" &&
        typeof ppexecCosts.priceSteps === "number")))
        return false;
    if (!((pp.maxTxExecutionUnits === undefined ||
        pp.maxTxExecutionUnits instanceof plutus_machine_1.ExBudget ||
        plutus_machine_1.ExBudget.isJson(pp.maxTxExecutionUnits)) &&
        (pp.maxBlockExecutionUnits === undefined ||
            pp.maxBlockExecutionUnits instanceof plutus_machine_1.ExBudget ||
            plutus_machine_1.ExBudget.isJson(pp.maxBlockExecutionUnits))))
        return false;
    if (!((pp.costModels === undefined || (0, cardano_costmodels_ts_1.isCostModels)(pp.costModels))))
        return false;
    return true;
}
exports.isPartialBabbageProtocolParameters = isPartialBabbageProtocolParameters;
function mapUIntEntryOrUndefined(tag, a) {
    return a === undefined ? undefined : {
        k: new cbor_1.CborUInt(tag),
        v: new cbor_1.CborUInt((0, ints_1.forceBigUInt)(a))
    };
}
function fromUIntOrUndef(n) {
    return n instanceof cbor_1.CborUInt ? n.num : undefined;
}
function kv(k, v) {
    return v === undefined ? undefined : {
        k: new cbor_1.CborUInt(k),
        v: v
    };
}
function partialBabbageProtocolParametersToCborObj(pps) {
    var protocolVersion = pps.protocolVersion, executionUnitPrices = pps.executionUnitPrices, maxTxExecutionUnits = pps.maxTxExecutionUnits, maxBlockExecutionUnits = pps.maxBlockExecutionUnits, costModels = pps.costModels;
    var costModelsKeys = Object.keys(costModels !== null && costModels !== void 0 ? costModels : {});
    return new cbor_1.CborMap([
        mapUIntEntryOrUndefined(0, pps.txFeePerByte),
        mapUIntEntryOrUndefined(1, pps.txFeeFixed),
        mapUIntEntryOrUndefined(2, pps.maxBlockBodySize),
        mapUIntEntryOrUndefined(3, pps.maxTxSize),
        mapUIntEntryOrUndefined(4, pps.maxBlockHeaderSize),
        mapUIntEntryOrUndefined(5, pps.stakeAddressDeposit),
        mapUIntEntryOrUndefined(6, pps.stakePoolDeposit),
        mapUIntEntryOrUndefined(7, pps.poolRetireMaxEpoch),
        mapUIntEntryOrUndefined(8, pps.stakePoolTargetNum),
        kv(9, (0, Rational_1.tryCborFromRational)(pps.poolPledgeInfluence)),
        kv(10, (0, Rational_1.tryCborFromRational)(pps.monetaryExpansion)),
        kv(11, (0, Rational_1.tryCborFromRational)(pps.treasuryCut)),
        protocolVersion === undefined ? undefined :
            kv(14, new protocolVersion_1.ProtocolVersion(protocolVersion).toCborObj()),
        mapUIntEntryOrUndefined(16, pps.minPoolCost),
        mapUIntEntryOrUndefined(17, pps.utxoCostPerByte),
        kv(18, (costModels === undefined ||
            (!costModelsKeys.includes("PlutusV1") && !costModelsKeys.includes("PlutusV2") && !costModelsKeys.includes("PlutusV3"))) ?
            undefined :
            (0, cardano_costmodels_ts_1.costModelsToCborObj)(costModels)),
        executionUnitPrices === undefined ? undefined :
            kv(19, Array.isArray(executionUnitPrices) ?
                new cbor_1.CborArray(executionUnitPrices) :
                new cbor_1.CborArray([
                    cbor_1.CborPositiveRational.fromNumber(executionUnitPrices.priceMemory),
                    cbor_1.CborPositiveRational.fromNumber(executionUnitPrices.priceSteps),
                ])),
        kv(20, plutus_machine_1.ExBudget.isJson(maxTxExecutionUnits) ? plutus_machine_1.ExBudget.fromJson(maxTxExecutionUnits).toCborObj() : maxTxExecutionUnits === null || maxTxExecutionUnits === void 0 ? void 0 : maxTxExecutionUnits.toCborObj()),
        kv(21, plutus_machine_1.ExBudget.isJson(maxBlockExecutionUnits) ? plutus_machine_1.ExBudget.fromJson(maxBlockExecutionUnits).toCborObj() : maxBlockExecutionUnits === null || maxBlockExecutionUnits === void 0 ? void 0 : maxBlockExecutionUnits.toCborObj()),
        mapUIntEntryOrUndefined(22, pps.maxValueSize),
        mapUIntEntryOrUndefined(23, pps.collateralPercentage),
        mapUIntEntryOrUndefined(24, pps.maxCollateralInputs)
    ].filter(function (elem) { return elem !== undefined; }));
}
exports.partialBabbageProtocolParametersToCborObj = partialBabbageProtocolParametersToCborObj;
function partialBabbageProtocolParametersToData(pps) {
    return cborToDataLitteral(partialBabbageProtocolParametersToCborObj(pps));
}
exports.partialBabbageProtocolParametersToData = partialBabbageProtocolParametersToData;
function cborToDataLitteral(cbor) {
    if (cbor instanceof cbor_1.CborPositiveRational) {
        return new plutus_data_1.DataConstr(0, [new plutus_data_1.DataI(cbor.num), new plutus_data_1.DataI(cbor.den)]);
    }
    if (cbor instanceof cbor_1.CborUInt || cbor instanceof cbor_1.CborNegInt) {
        return new plutus_data_1.DataI(cbor.num);
    }
    if (cbor instanceof cbor_1.CborBytes) {
        return new plutus_data_1.DataB(cbor.bytes);
    }
    if (cbor instanceof cbor_1.CborText) {
        return new plutus_data_1.DataB((0, uint8array_utils_1.fromUtf8)(cbor.text));
    }
    if (cbor instanceof cbor_1.CborArray) {
        return new plutus_data_1.DataList(cbor.array.map(cborToDataLitteral));
    }
    if (cbor instanceof cbor_1.CborMap) {
        return new plutus_data_1.DataMap(cbor.map.map(function (_a) {
            var k = _a.k, v = _a.v;
            return new plutus_data_1.DataPair(cborToDataLitteral(k), cborToDataLitteral(v));
        }));
    }
    if (cbor instanceof cbor_1.CborTag) {
        return cborToDataLitteral(cbor.data);
    }
    throw new Error("unsupported cbor for literal data");
}
var maxProtocolParamsEntries = 25;
function partialBabbageProtocolParametersFromCborObj(cObj) {
    var _a;
    if (!(cObj instanceof cbor_1.CborMap))
        throw new Error("Invalid CBOR format for \"Partial<BabbageProtocolParameters>\"");
    var fields = new Array(maxProtocolParamsEntries).fill(undefined);
    var _loop_1 = function (i) {
        var v = ((_a = cObj.map.find(function (_a) {
            var k = _a.k;
            return k instanceof cbor_1.CborUInt && Number(k.num) === i;
        })) !== null && _a !== void 0 ? _a : { v: undefined }).v;
        if (v === undefined)
            return "continue";
        fields[i] = v;
    };
    for (var i = 0; i <= maxProtocolParamsEntries; i++) {
        _loop_1(i);
    }
    var _b = __read(fields, 25), _minFeeCoeff = _b[0], // 0: txFeePerByte
    _minFeeFix = _b[1], // 1: txFeeFixed
    _maxBlockBodySize = _b[2], // 2: maxBlockBodySize
    _maxTxSize = _b[3], // 3: maxTxSize
    _maxBlockHeaderSize = _b[4], // 4: maxBlockHeaderSize
    _keyDep = _b[5], // 5: stakeAddressDeposit
    _poolDep = _b[6], // 6: stakePoolDeposit
    _epoch = _b[7], // 7: poolRetireMaxEpoch
    _kParam = _b[8], // 8: stakePoolTargetNum
    _pledgeInfluence = _b[9], // 9: poolPledgeInfluence
    _expansionRate = _b[10], // 10: monetaryExpansion
    _treasureryGrowthRate = _b[11], // 11: treasuryCut
    _12 = _b[12], // 12: decentralization constant (Shelley, removed in later eras)
    _13 = _b[13], // 13: extra entropy (Shelley, removed in later eras)
    _protocolVersion = _b[14], // 14: protocolVersion
    _15 = _b[15], // 15: minUTxOValue (Shelley/Mary, replaced by utxoCostPerByte)
    _poolMinFee = _b[16], // 16: minPoolCost
    _adaPerUtxoByte = _b[17], // 17: utxoCostPerByte (Mary/Alonzo, aka lovelacePerUTxOWord in Babbage)
    _costmdls = _b[18], // 18: costModels (Alonzo/Babbage)
    _execCosts = _b[19], // 19: executionUnitPrices (Alonzo/Babbage)
    _maxTxExecUnits = _b[20], // 20: maxTxExecutionUnits (Alonzo/Babbage)
    _maxBlockExecUnits = _b[21], // 21: maxBlockExecutionUnits (Alonzo/Babbage)
    _maxValueSize = _b[22], // 22: maxValueSize (Mary/Alonzo/Babbage)
    _collatearalPerc = _b[23], // 23: collateralPercentage (Alonzo/Babbage)
    _maxCollIns = _b[24];
    var protocolVersion;
    try {
        protocolVersion = _protocolVersion ? protocolVersion_1.ProtocolVersion.fromCborObj(_protocolVersion) : undefined;
    }
    catch (_c) { }
    var executionUnitPrices = undefined;
    if (_execCosts instanceof cbor_1.CborArray) {
        var mem_price = cbor_1.CborPositiveRational.fromCborObjOrUndef(_execCosts.array[0]);
        var cpu_price = cbor_1.CborPositiveRational.fromCborObjOrUndef(_execCosts.array[1]);
        executionUnitPrices = mem_price !== undefined && cpu_price !== undefined ? [mem_price, cpu_price] : undefined;
    }
    var _costModels = (0, cardano_costmodels_ts_1.costModelsFromCborObj)(_costmdls);
    return {
        txFeePerByte: fromUIntOrUndef(_minFeeCoeff),
        txFeeFixed: fromUIntOrUndef(_minFeeFix),
        maxBlockBodySize: fromUIntOrUndef(_maxBlockBodySize),
        maxTxSize: fromUIntOrUndef(_maxTxSize),
        maxBlockHeaderSize: fromUIntOrUndef(_maxBlockHeaderSize),
        stakeAddressDeposit: fromUIntOrUndef(_keyDep),
        stakePoolDeposit: fromUIntOrUndef(_poolDep),
        poolRetireMaxEpoch: fromUIntOrUndef(_epoch),
        stakePoolTargetNum: fromUIntOrUndef(_kParam),
        poolPledgeInfluence: cbor_1.CborPositiveRational.fromCborObjOrUndef(_pledgeInfluence),
        monetaryExpansion: cbor_1.CborPositiveRational.fromCborObjOrUndef(_expansionRate),
        treasuryCut: cbor_1.CborPositiveRational.fromCborObjOrUndef(_treasureryGrowthRate),
        protocolVersion: protocolVersion,
        minPoolCost: fromUIntOrUndef(_poolMinFee),
        utxoCostPerByte: fromUIntOrUndef(_adaPerUtxoByte),
        costModels: Object.keys(_costModels).length === 0 ? undefined : _costModels,
        executionUnitPrices: executionUnitPrices,
        maxTxExecutionUnits: _maxTxExecUnits === undefined ? undefined : plutus_machine_1.ExBudget.fromCborObj(_maxTxExecUnits),
        maxBlockExecutionUnits: _maxBlockExecUnits === undefined ? undefined : plutus_machine_1.ExBudget.fromCborObj(_maxBlockExecUnits),
        maxValueSize: fromUIntOrUndef(_maxValueSize),
        collateralPercentage: fromUIntOrUndef(_collatearalPerc),
        maxCollateralInputs: fromUIntOrUndef(_maxCollIns)
    };
}
exports.partialBabbageProtocolParametersFromCborObj = partialBabbageProtocolParametersFromCborObj;
exports.defaultBabbageProtocolParameters = (0, obj_utils_1.freezeAll)({
    txFeePerByte: 44,
    txFeeFixed: 155381,
    maxBlockBodySize: 90112,
    maxTxSize: 16384,
    maxBlockHeaderSize: 1100,
    stakeAddressDeposit: 2000000,
    stakePoolDeposit: 500000000,
    poolRetireMaxEpoch: 18,
    stakePoolTargetNum: 500,
    poolPledgeInfluence: new cbor_1.CborPositiveRational(3, 10),
    monetaryExpansion: new cbor_1.CborPositiveRational(3, 1000),
    treasuryCut: new cbor_1.CborPositiveRational(2, 10),
    protocolVersion: new protocolVersion_1.ProtocolVersion({ major: 7, minor: 0 }),
    minPoolCost: 340000000,
    utxoCostPerByte: 4310,
    costModels: {
        PlutusScriptV1: cardano_costmodels_ts_1.defaultV1Costs,
        PlutusScriptV2: cardano_costmodels_ts_1.defaultV2Costs, // Plutus V2 cost model introduced in Babbage/Vasil
    },
    executionUnitPrices: [
        cbor_1.CborPositiveRational.fromNumber(0.0577),
        cbor_1.CborPositiveRational.fromNumber(0.0000721) // CPU/Steps price (historical approximation, may be adjusted)
    ],
    maxTxExecutionUnits: new plutus_machine_1.ExBudget({ mem: 14000000, cpu: 11500000000 }),
    maxBlockExecutionUnits: new plutus_machine_1.ExBudget({ mem: 62000000, cpu: 50000000000 }),
    maxValueSize: 5000,
    collateralPercentage: 150,
    maxCollateralInputs: 3 // Unchanged, maximum number of collateral inputs
});
function partialBabbageProtocolParamsToJson(pp) {
    var _a, _b, _c, _d, _e;
    return __assign(__assign({}, pp), { poolPledgeInfluence: typeof pp.poolPledgeInfluence === "number" ? pp.poolPledgeInfluence : (_a = pp.poolPledgeInfluence) === null || _a === void 0 ? void 0 : _a.toNumber(), monetaryExpansion: typeof pp.monetaryExpansion === "number" ? pp.monetaryExpansion : (_b = pp.monetaryExpansion) === null || _b === void 0 ? void 0 : _b.toNumber(), treasuryCut: typeof pp.treasuryCut === "number" ? pp.treasuryCut : (_c = pp.treasuryCut) === null || _c === void 0 ? void 0 : _c.toNumber(), costModels: pp.costModels === undefined ? undefined : (0, cardano_costmodels_ts_1.costModelsToJson)(pp.costModels), executionUnitPrices: Array.isArray(pp.executionUnitPrices) ?
            {
                priceSteps: pp.executionUnitPrices[1].toNumber(),
                priceMemory: pp.executionUnitPrices[0].toNumber()
            } :
            pp.executionUnitPrices, maxTxExecutionUnits: plutus_machine_1.ExBudget.isJson(pp.maxTxExecutionUnits) ? pp.maxTxExecutionUnits : (_d = pp.maxTxExecutionUnits) === null || _d === void 0 ? void 0 : _d.toJson(), maxBlockExecutionUnits: plutus_machine_1.ExBudget.isJson(pp.maxBlockExecutionUnits) ? pp.maxBlockExecutionUnits : (_e = pp.maxBlockExecutionUnits) === null || _e === void 0 ? void 0 : _e.toJson() });
}
exports.partialBabbageProtocolParamsToJson = partialBabbageProtocolParamsToJson;
