// @harmoniclabs/cardano-ledger-ts/dist/eras/shelley/tx/ShelleyUTxO.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxOutRef = exports.ShelleyUTxO = exports.isIShelleyUTxO = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var InvalidCborFormatError_1 = require("../../../utils/InvalidCborFormatError.js");
var ShelleyTxOut_1 = require("./ShelleyTxOut.js");
var TxOutRef_1 = require("../../common/TxOutRef.js");
Object.defineProperty(exports, "TxOutRef", { enumerable: true, get: function () { return TxOutRef_1.TxOutRef; } });
var getSubCborRef_1 = require("../../../utils/getSubCborRef.js");
var HasUtxoRef_1 = require("../../common/HasUtxoRef.js");
function isIShelleyUTxO(stuff) {
    return ((0, obj_utils_1.isObject)(stuff)
        // && hasOwn( stuff, "utxoRef" ) && isITxOutRef( stuff.utxoRef )
        && (0, HasUtxoRef_1.hasUtxoRef)(stuff)
        && (0, obj_utils_1.hasOwn)(stuff, "resolved") && (0, ShelleyTxOut_1.isIShelleyTxOut)(stuff.resolved));
}
exports.isIShelleyUTxO = isIShelleyUTxO;
var ShelleyUTxO = /** @class */ (function () {
    function ShelleyUTxO(_a, cborRef) {
        var utxoRef = _a.utxoRef, resolved = _a.resolved;
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        this.utxoRef = utxoRef instanceof TxOutRef_1.TxOutRef ? utxoRef : new TxOutRef_1.TxOutRef(utxoRef);
        this.resolved = resolved instanceof ShelleyTxOut_1.ShelleyTxOut ? resolved : new ShelleyTxOut_1.ShelleyTxOut(resolved);
    }
    ShelleyUTxO.prototype.clone = function () {
        return new ShelleyUTxO(this);
    };
    ShelleyUTxO.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    ShelleyUTxO.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.cborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    ShelleyUTxO.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        }
        return new cbor_1.CborArray([
            this.utxoRef.toCborObj(),
            this.resolved.toCborObj()
        ]);
    };
    ShelleyUTxO.fromCbor = function (cStr) {
        return ShelleyUTxO.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cStr), { keepRef: true }));
    };
    ShelleyUTxO.fromCborObj = function (cObj) {
        if (!(cObj instanceof cbor_1.CborArray
            && cObj.array.length >= 2))
            throw new InvalidCborFormatError_1.InvalidCborFormatError("ShelleyUTxO");
        var _a = __read(cObj.array, 2), ref = _a[0], res = _a[1];
        var utxoRef;
        var resolved;
        if (ref === undefined)
            throw new InvalidCborFormatError_1.InvalidCborFormatError("ShelleyUTxO");
        if (res === undefined)
            throw new InvalidCborFormatError_1.InvalidCborFormatError("ShelleyUTxO", "if you are trying to parse only a TxOutRef instead (<hex>#<index>) you should use `TxOutRef.fromCborObj`");
        utxoRef = TxOutRef_1.TxOutRef.fromCborObj(ref);
        resolved = ShelleyTxOut_1.ShelleyTxOut.fromCborObj(res);
        return new ShelleyUTxO({
            utxoRef: utxoRef,
            resolved: resolved
        }, (0, getSubCborRef_1.getSubCborRef)(cObj));
    };
    ShelleyUTxO.prototype.toJSON = function () { return this.toJson(); };
    ShelleyUTxO.prototype.toJson = function () {
        return {
            utxoRef: this.utxoRef.toJson(),
            resolved: this.resolved.toJson()
        };
    };
    ShelleyUTxO.sort = function (a, b) {
        return TxOutRef_1.TxOutRef.sort(a.utxoRef, b.utxoRef);
    };
    return ShelleyUTxO;
}());
exports.ShelleyUTxO = ShelleyUTxO;
