// @harmoniclabs/cardano-ledger-ts/dist/eras/allegra/protocol/AllegraProtocolParameters.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.partialAllegraProtocolParamsToJson = exports.defaultAllegraProtocolParameters = exports.partialAllegraProtocolParametersFromCborObj = exports.partialAllegraProtocolParametersToData = exports.partialAllegraProtocolParametersToCborObj = exports.isPartialAllegraProtocolParameters = exports.isAllegraProtocolParameters = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var uint8array_utils_1 = require("@harmoniclabs/uint8array-utils");
var ints_1 = require("../../../utils/ints.js");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var common_1 = require("../../common/index.js");
function isAllegraProtocolParameters(something) {
    var expectedKeys = [
        "txFeePerByte",
        "txFeeFixed",
        "maxBlockBodySize",
        "maxTxSize",
        "maxBlockHeaderSize",
        "stakeAddressDeposit",
        "stakePoolDeposit",
        "poolRetireMaxEpoch",
        "stakePoolTargetNum",
        "poolPledgeInfluence",
        "monetaryExpansion",
        "treasuryCut",
        "protocolVersion",
        "minPoolCost"
    ];
    var actualKeys = Object.keys(something);
    if (!expectedKeys.every(function (k) { return actualKeys.includes(k); }))
        return false;
    var pp = something;
    if (![
        "txFeePerByte",
        "txFeeFixed",
        "maxBlockBodySize",
        "maxTxSize",
        "maxBlockHeaderSize",
        "stakeAddressDeposit",
        "stakePoolDeposit",
        "poolRetireMaxEpoch",
        "stakePoolTargetNum",
        "minPoolCost"
    ].every(function (uintKey) { return (0, ints_1.canBeUInteger)(pp[uintKey]); }))
        return false;
    if (!((0, common_1.isRational)(pp.poolPledgeInfluence) &&
        (0, common_1.isRational)(pp.monetaryExpansion) &&
        (0, common_1.isRational)(pp.treasuryCut)))
        return false;
    var ppv = pp.protocolVersion;
    if (!(
    // protocolVersion removed in conway
    ppv === undefined ||
        (0, common_1.isIProtocolVersion)(ppv)))
        return false;
    return true;
}
exports.isAllegraProtocolParameters = isAllegraProtocolParameters;
function isPartialAllegraProtocolParameters(something) {
    if (!(0, obj_utils_1.isObject)(something))
        return false;
    var pp = something;
    if (![
        "txFeePerByte",
        "txFeeFixed",
        "maxBlockBodySize",
        "maxTxSize",
        "maxBlockHeaderSize",
        "stakeAddressDeposit",
        "stakePoolDeposit",
        "poolRetireMaxEpoch",
        "stakePoolTargetNum",
        "minPoolCost"
    ].every(function (uintKey) { return pp[uintKey] === undefined || (0, ints_1.canBeUInteger)(pp[uintKey]); }))
        return false;
    if (!((0, common_1.isRationalOrUndefined)(pp.poolPledgeInfluence) &&
        (0, common_1.isRationalOrUndefined)(pp.monetaryExpansion) &&
        (0, common_1.isRationalOrUndefined)(pp.treasuryCut)))
        return false;
    var ppv = pp.protocolVersion;
    if (!(ppv === undefined ||
        (0, common_1.isIProtocolVersion)(ppv)))
        return false;
    return true;
}
exports.isPartialAllegraProtocolParameters = isPartialAllegraProtocolParameters;
function mapUIntEntryOrUndefined(tag, a) {
    return a === undefined ? undefined : {
        k: new cbor_1.CborUInt(tag),
        v: new cbor_1.CborUInt((0, ints_1.forceBigUInt)(a))
    };
}
function fromUIntOrUndef(n) {
    return n instanceof cbor_1.CborUInt ? n.num : undefined;
}
function kv(k, v) {
    return v === undefined ? undefined : {
        k: new cbor_1.CborUInt(k),
        v: v
    };
}
function partialAllegraProtocolParametersToCborObj(pps) {
    var protocolVersion = pps.protocolVersion;
    return new cbor_1.CborMap([
        mapUIntEntryOrUndefined(0, pps.txFeePerByte),
        mapUIntEntryOrUndefined(1, pps.txFeeFixed),
        mapUIntEntryOrUndefined(2, pps.maxBlockBodySize),
        mapUIntEntryOrUndefined(3, pps.maxTxSize),
        mapUIntEntryOrUndefined(4, pps.maxBlockHeaderSize),
        mapUIntEntryOrUndefined(5, pps.stakeAddressDeposit),
        mapUIntEntryOrUndefined(6, pps.stakePoolDeposit),
        mapUIntEntryOrUndefined(7, pps.poolRetireMaxEpoch),
        mapUIntEntryOrUndefined(8, pps.stakePoolTargetNum),
        kv(9, (0, common_1.tryCborFromRational)(pps.poolPledgeInfluence)),
        kv(10, (0, common_1.tryCborFromRational)(pps.monetaryExpansion)),
        kv(11, (0, common_1.tryCborFromRational)(pps.treasuryCut)),
        protocolVersion === undefined ? undefined :
            kv(14, new common_1.ProtocolVersion(protocolVersion).toCborObj()),
        mapUIntEntryOrUndefined(16, pps.minPoolCost),
    ].filter(function (elem) { return elem !== undefined; }));
}
exports.partialAllegraProtocolParametersToCborObj = partialAllegraProtocolParametersToCborObj;
function partialAllegraProtocolParametersToData(pps) {
    return cborToDataLitteral(partialAllegraProtocolParametersToCborObj(pps));
}
exports.partialAllegraProtocolParametersToData = partialAllegraProtocolParametersToData;
function cborToDataLitteral(cbor) {
    if (cbor instanceof cbor_1.CborPositiveRational) {
        return new plutus_data_1.DataConstr(0, [new plutus_data_1.DataI(cbor.num), new plutus_data_1.DataI(cbor.den)]);
    }
    if (cbor instanceof cbor_1.CborUInt || cbor instanceof cbor_1.CborNegInt) {
        return new plutus_data_1.DataI(cbor.num);
    }
    if (cbor instanceof cbor_1.CborBytes) {
        return new plutus_data_1.DataB(cbor.bytes);
    }
    if (cbor instanceof cbor_1.CborText) {
        return new plutus_data_1.DataB((0, uint8array_utils_1.fromUtf8)(cbor.text));
    }
    if (cbor instanceof cbor_1.CborArray) {
        return new plutus_data_1.DataList(cbor.array.map(cborToDataLitteral));
    }
    if (cbor instanceof cbor_1.CborMap) {
        return new plutus_data_1.DataMap(cbor.map.map(function (_a) {
            var k = _a.k, v = _a.v;
            return new plutus_data_1.DataPair(cborToDataLitteral(k), cborToDataLitteral(v));
        }));
    }
    if (cbor instanceof cbor_1.CborTag) {
        return cborToDataLitteral(cbor.data);
    }
    throw new Error("unsupported cbor for litteral data");
}
var maxProtocolParamsEntries = 17;
function partialAllegraProtocolParametersFromCborObj(cObj) {
    var _a;
    if (!(cObj instanceof cbor_1.CborMap))
        throw new Error("Invalid CBOR format for \"Partial<AllegraProtocolParameters>\"");
    var fields = new Array(maxProtocolParamsEntries).fill(undefined);
    var _loop_1 = function (i) {
        var v = ((_a = cObj.map.find(function (_a) {
            var k = _a.k;
            return k instanceof cbor_1.CborUInt && Number(k.num) === i;
        })) !== null && _a !== void 0 ? _a : { v: undefined }).v;
        if (v === undefined)
            return "continue";
        fields[i] = v;
    };
    for (var i = 0; i <= maxProtocolParamsEntries; i++) {
        _loop_1(i);
    }
    var _b = __read(fields, 17), _minFeeCoeff = _b[0], // 0: txFeePerByte
    _minFeeFix = _b[1], // 1: txFeeFixed
    _maxBlockBodySize = _b[2], // 2: maxBlockBodySize
    _maxTxSize = _b[3], // 3: maxTxSize
    _maxBlockHeaderSize = _b[4], // 4: maxBlockHeaderSize
    _keyDep = _b[5], // 5: stakeAddressDeposit
    _poolDep = _b[6], // 6: stakePoolDeposit
    _epoch = _b[7], // 7: poolRetireMaxEpoch
    _kParam = _b[8], // 8: stakePoolTargetNum
    _pledgeInfluence = _b[9], // 9: poolPledgeInfluence
    _expansionRate = _b[10], // 10: monetaryExpansion
    _treasureryGrowthRate = _b[11], // 11: treasuryCut
    _12 = _b[12], // 12: decentralization constant (Shelley, removed in later eras)
    _13 = _b[13], // 13: extra entropy (Shelley, removed in later eras)
    _protocolVersion = _b[14], // 14: protocolVersion
    _15 = _b[15], // 15: minUTxOValue (Shelley/Mary, replaced by utxoCostPerByte)
    _poolMinFee = _b[16];
    var protocolVersion;
    try {
        protocolVersion = _protocolVersion ? common_1.ProtocolVersion.fromCborObj(_protocolVersion) : undefined;
    }
    catch (_c) { }
    return {
        txFeePerByte: fromUIntOrUndef(_minFeeCoeff),
        txFeeFixed: fromUIntOrUndef(_minFeeFix),
        maxBlockBodySize: fromUIntOrUndef(_maxBlockBodySize),
        maxTxSize: fromUIntOrUndef(_maxTxSize),
        maxBlockHeaderSize: fromUIntOrUndef(_maxBlockHeaderSize),
        stakeAddressDeposit: fromUIntOrUndef(_keyDep),
        stakePoolDeposit: fromUIntOrUndef(_poolDep),
        poolRetireMaxEpoch: fromUIntOrUndef(_epoch),
        stakePoolTargetNum: fromUIntOrUndef(_kParam),
        poolPledgeInfluence: cbor_1.CborPositiveRational.fromCborObjOrUndef(_pledgeInfluence),
        monetaryExpansion: cbor_1.CborPositiveRational.fromCborObjOrUndef(_expansionRate),
        treasuryCut: cbor_1.CborPositiveRational.fromCborObjOrUndef(_treasureryGrowthRate),
        protocolVersion: protocolVersion,
        minPoolCost: fromUIntOrUndef(_poolMinFee)
    };
}
exports.partialAllegraProtocolParametersFromCborObj = partialAllegraProtocolParametersFromCborObj;
exports.defaultAllegraProtocolParameters = (0, obj_utils_1.freezeAll)({
    txFeePerByte: 44,
    txFeeFixed: 155381,
    maxBlockBodySize: 65536,
    maxTxSize: 16384,
    maxBlockHeaderSize: 1100,
    stakeAddressDeposit: 2000000,
    stakePoolDeposit: 500000000,
    poolRetireMaxEpoch: 18,
    stakePoolTargetNum: 150,
    poolPledgeInfluence: new cbor_1.CborPositiveRational(3, 10),
    monetaryExpansion: new cbor_1.CborPositiveRational(3, 1000),
    treasuryCut: new cbor_1.CborPositiveRational(2, 10),
    protocolVersion: new common_1.ProtocolVersion({ major: 2, minor: 0 }),
    minPoolCost: 340000000
});
function partialAllegraProtocolParamsToJson(pp) {
    var _a, _b, _c;
    return __assign(__assign({}, pp), { poolPledgeInfluence: typeof pp.poolPledgeInfluence === "number" ? pp.poolPledgeInfluence : (_a = pp.poolPledgeInfluence) === null || _a === void 0 ? void 0 : _a.toNumber(), monetaryExpansion: typeof pp.monetaryExpansion === "number" ? pp.monetaryExpansion : (_b = pp.monetaryExpansion) === null || _b === void 0 ? void 0 : _b.toNumber(), treasuryCut: typeof pp.treasuryCut === "number" ? pp.treasuryCut : (_c = pp.treasuryCut) === null || _c === void 0 ? void 0 : _c.toNumber() });
}
exports.partialAllegraProtocolParamsToJson = partialAllegraProtocolParamsToJson;
