// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/keep-alive/messages/KeepAliveDone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeepAliveDone = void 0;
exports.isIKeepAliveDone = isIKeepAliveDone;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isIKeepAliveDone(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var KeepAliveDone = /** @class */ (function () {
    function KeepAliveDone() {
    }
    ;
    KeepAliveDone.prototype.toJSON = function () { return this.toJson(); };
    KeepAliveDone.prototype.toJson = function () { return {}; };
    KeepAliveDone.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    KeepAliveDone.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    KeepAliveDone.prototype.toCborObj = function () {
        return new cbor_1.CborArray([new cbor_1.CborUInt(2)]);
    };
    KeepAliveDone.fromCbor = function (cbor) {
        var buff = cbor instanceof Uint8Array ?
            cbor :
            (0, cbor_1.forceCborString)(cbor).toBuffer();
        return KeepAliveDone.fromCborObj(cbor_1.Cbor.parse(buff, { keepRef: false }));
    };
    KeepAliveDone.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(2)))
            throw new Error("invalid CBOR for 'KeepAliveDone");
        return new KeepAliveDone();
    };
    return KeepAliveDone;
}());
exports.KeepAliveDone = KeepAliveDone;
