// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/keep-alive/messages/KeepAliveResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeepAliveResponse = void 0;
exports.isIKeepAliveResponse = isIKeepAliveResponse;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var isWord16_1 = require("../../utils/isWord16.js");
function isIKeepAliveResponse(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var KeepAliveResponse = /** @class */ (function () {
    function KeepAliveResponse(_a) {
        var cookie = _a.cookie;
        if (!(0, isWord16_1.isWord16)(cookie)) {
            throw new Error("keep alive cookie is not word 16");
        }
        this.cookie = Number(cookie) | 0;
    }
    KeepAliveResponse.prototype.toJSON = function () { return this.toJson(); };
    KeepAliveResponse.prototype.toJson = function () { return {}; };
    KeepAliveResponse.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    KeepAliveResponse.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    KeepAliveResponse.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(1),
            new cbor_1.CborUInt(this.cookie)
        ]);
    };
    KeepAliveResponse.fromCbor = function (cbor) {
        var buff = cbor instanceof Uint8Array ?
            cbor :
            (0, cbor_1.forceCborString)(cbor).toBuffer();
        return KeepAliveResponse.fromCborObj(cbor_1.Cbor.parse(buff));
    };
    KeepAliveResponse.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[1] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(1)))
            throw new Error("invalid CBOR for 'KeepAliveResponse");
        return new KeepAliveResponse({
            cookie: cbor.array[1].num
        });
    };
    return KeepAliveResponse;
}());
exports.KeepAliveResponse = KeepAliveResponse;
