// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/keep-alive/messages/KeepAliveRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeepAliveRequest = void 0;
exports.isIKeepAliveRequest = isIKeepAliveRequest;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var isWord16_1 = require("../../utils/isWord16.js");
function isIKeepAliveRequest(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var KeepAliveRequest = /** @class */ (function () {
    function KeepAliveRequest(_a) {
        var cookie = _a.cookie;
        if (!(0, isWord16_1.isWord16)(cookie)) {
            throw new Error("keep alive cookie is not word 16");
        }
        this.cookie = Number(cookie) | 0;
    }
    KeepAliveRequest.prototype.toJSON = function () { return this.toJson(); };
    KeepAliveRequest.prototype.toJson = function () { return {}; };
    KeepAliveRequest.prototype.toCborBytes = function () {
        return this.toCbor().toBuffer();
    };
    KeepAliveRequest.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    KeepAliveRequest.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(0),
            new cbor_1.CborUInt(this.cookie)
        ]);
    };
    KeepAliveRequest.fromCbor = function (cbor) {
        var buff = cbor instanceof Uint8Array ?
            cbor :
            (0, cbor_1.forceCborString)(cbor).toBuffer();
        return KeepAliveRequest.fromCborObj(cbor_1.Cbor.parse(buff, { keepRef: false }));
    };
    KeepAliveRequest.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[1] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(0)))
            throw new Error("invalid CBOR for 'KeepAliveRequest");
        return new KeepAliveRequest({
            cookie: cbor.array[1].num
        });
    };
    return KeepAliveRequest;
}());
exports.KeepAliveRequest = KeepAliveRequest;
