// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/chain-sync/messages/ChainSyncRequestNext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainSyncRequestNext = void 0;
exports.isIChainSyncRequestNext = isIChainSyncRequestNext;
var cbor_1 = require("@harmoniclabs/cbor");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
function isIChainSyncRequestNext(stuff) {
    return (0, obj_utils_1.isObject)(stuff);
}
var ChainSyncRequestNext = /** @class */ (function () {
    function ChainSyncRequestNext() {
        this.cborRef = undefined;
    }
    ;
    ChainSyncRequestNext.prototype.toJSON = function () { return this.toJson(); };
    ChainSyncRequestNext.prototype.toJson = function () { return {}; };
    ChainSyncRequestNext.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    ChainSyncRequestNext.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    ChainSyncRequestNext.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([new cbor_1.CborUInt(0)]);
    };
    ChainSyncRequestNext.fromCbor = function (cbor) {
        var buff = cbor instanceof Uint8Array ?
            cbor :
            (0, cbor_1.forceCborString)(cbor).toBuffer();
        return ChainSyncRequestNext.fromCborObj(cbor_1.Cbor.parse(buff), buff);
    };
    ChainSyncRequestNext.fromCborObj = function (cbor, originalBytes) {
        if (originalBytes === void 0) { originalBytes = undefined; }
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(0)))
            throw new Error("invalid CBOR for 'ChainSyncRequestNext");
        return new ChainSyncRequestNext();
    };
    return ChainSyncRequestNext;
}());
exports.ChainSyncRequestNext = ChainSyncRequestNext;
