// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/chain-sync/messages/ChainSyncRollBackwards.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainSyncRollBackwards = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var ChainPoint_1 = require("../../types/ChainPoint.js");
var ChainTip_1 = require("../../types/ChainTip.js");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
var ChainSyncRollBackwards = /** @class */ (function () {
    function ChainSyncRollBackwards(rollback, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        var point = rollback.point, tip = rollback.tip;
        if (!((0, ChainPoint_1.isIChainPoint)(point) &&
            (0, ChainTip_1.isIChainTip)(tip)))
            throw new Error("invalid IChainSyncRollBackwards interface");
        this.point = point instanceof ChainPoint_1.ChainPoint ? point : new ChainPoint_1.ChainPoint(point);
        this.tip = tip instanceof ChainTip_1.ChainTip ? tip : new ChainTip_1.ChainTip(tip);
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(rollback);
    }
    ;
    ChainSyncRollBackwards.prototype.toJSON = function () { return this.toJson(); };
    ChainSyncRollBackwards.prototype.toJson = function () {
        return {
            point: this.point.toJson(),
            tip: this.tip.toJson()
        };
    };
    ChainSyncRollBackwards.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    ChainSyncRollBackwards.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    ChainSyncRollBackwards.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(3),
            this.point.toCborObj(),
            this.tip.toCborObj()
        ]);
    };
    ChainSyncRollBackwards.fromCbor = function (cbor) {
        var buff = cbor instanceof Uint8Array ?
            cbor :
            (0, cbor_1.forceCborString)(cbor).toBuffer();
        return ChainSyncRollBackwards.fromCborObj(cbor_1.Cbor.parse(buff), buff);
    };
    ChainSyncRollBackwards.fromCborObj = function (cbor, originalBytes) {
        if (originalBytes === void 0) { originalBytes = undefined; }
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 3 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(3)))
            throw new Error("invalid CBOR for 'ChainSyncRollBackwards");
        var _a = __read(cbor.array, 3), _idx = _a[0], pointCbor = _a[1], tipCbor = _a[2];
        return new ChainSyncRollBackwards({
            point: ChainPoint_1.ChainPoint.fromCborObj(pointCbor),
            tip: ChainTip_1.ChainTip.fromCborObj(tipCbor)
        }, (0, getSubCborRef_1.getSubCborRef)(cbor, originalBytes));
    };
    return ChainSyncRollBackwards;
}());
exports.ChainSyncRollBackwards = ChainSyncRollBackwards;
