// @harmoniclabs/ouroboros-miniprotocols-ts/dist/protocols/chain-sync/messages/ChainSyncRollForward.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainSyncRollForward = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var ChainTip_1 = require("../../types/ChainTip.js");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
var uint8array_utils_1 = require("@harmoniclabs/uint8array-utils");
var ChainSyncRollForward = /** @class */ (function () {
    function ChainSyncRollForward(forward, cborRef) {
        if (cborRef === void 0) { cborRef = undefined; }
        this.cborRef = cborRef;
        var data = forward.data, tip = forward.tip;
        if (!((0, cbor_1.isCborObj)(data) &&
            (0, ChainTip_1.isIChainTip)(tip)))
            throw new Error("invalid IChainSyncRollForward interface");
        this.data = data;
        this.tip = tip instanceof ChainTip_1.ChainTip ? tip : new ChainTip_1.ChainTip(tip);
        this.cborRef = cborRef !== null && cborRef !== void 0 ? cborRef : (0, getSubCborRef_1.subCborRefOrUndef)(forward);
    }
    ;
    ChainSyncRollForward.prototype.toJSON = function () { return this.toJson(); };
    ChainSyncRollForward.prototype.toJson = function () {
        return {
            data: (0, uint8array_utils_1.toHex)(this.toCborBytes()),
            tip: this.tip.toJson()
        };
    };
    ChainSyncRollForward.prototype.toString = function () {
        return "(roll forward: (header: ".concat(cbor_1.Cbor.encode(this.data).toString(), ") ").concat(this.tip.toString(), " )");
    };
    ChainSyncRollForward.prototype.toCborBytes = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return this.cborRef.toBuffer();
        return this.toCbor().toBuffer();
    };
    ChainSyncRollForward.prototype.toCbor = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return new cbor_1.CborString(this.cborRef.toBuffer());
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    ChainSyncRollForward.prototype.toCborObj = function () {
        if (this.cborRef instanceof cbor_1.SubCborRef)
            return cbor_1.Cbor.parse(this.cborRef.toBuffer());
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(2),
            this.data,
            this.tip.toCborObj()
        ]);
    };
    /**
     * @returns {Uint8Array}
     * the bytes of `this.data` as present on `this.cborBytes`
     * (using `Cbor.parseLazy`)
     */
    ChainSyncRollForward.prototype.getDataBytes = function () {
        var msgData = this.toCborBytes();
        var lazy = cbor_1.Cbor.parseLazy(msgData);
        return lazy.array[1];
    };
    ChainSyncRollForward.fromCbor = function (cbor) {
        var buff = cbor instanceof Uint8Array ?
            cbor :
            (0, cbor_1.forceCborString)(cbor).toBuffer();
        return ChainSyncRollForward.fromCborObj(cbor_1.Cbor.parse(buff), buff);
    };
    ChainSyncRollForward.fromCborObj = function (cbor, originalBytes) {
        if (originalBytes === void 0) { originalBytes = undefined; }
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 3 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            cbor.array[0].num === BigInt(2)))
            throw new Error("invalid CBOR for 'ChainSyncRollForward");
        var _a = __read(cbor.array, 3), _idx = _a[0], headerCbor = _a[1], tipCbor = _a[2];
        return new ChainSyncRollForward({
            data: headerCbor,
            tip: ChainTip_1.ChainTip.fromCborObj(tipCbor)
        }, (0, getSubCborRef_1.getSubCborRef)(cbor, originalBytes));
    };
    return ChainSyncRollForward;
}());
exports.ChainSyncRollForward = ChainSyncRollForward;
